In [1]:
import pandas as pd
import datetime

# 유튜브 인기동영상 데이터

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/youtube.csv",index_col=0)

In [3]:
df.head()

title             channelTitle  \
0                                          [신병] 물자창고                      장삐쭈   
1   RAIN(비) - 나로 바꾸자 Switch to me (duet with JYP) MV  RAIN's Official Channel   
2       2020년 제야의 종 온라인 타종행사 | 보신각 현장 행사는 진행하지 않습니다.              서울시 · Seoul   
3                                       고기남자의 칠면조 파티             고기남자 MeatMan   
4  골목 3mc를 분노하게 만든 마음고생이 심했을 공릉 백반집 사장님의 푸념?! [예능...                    스브스밥집   

   categoryId  view_count  likes  dislikes  comment_count  \
0          23     1893473  38249       730           8595   
1          10     2600864      0         0          20129   
2          29      347049   3564       120            178   
3          26      528458  15372       280           3470   
4          24      494904   3918       111           3142   

                  channelId trending_date2  
0  UChbE5OZQ6dRHECsX0tEPEZQ     2021-01-01  
1  UCxXgIeE5hxWxHG6dz9Scg2w     2021-01-01  
2  UCZUPZW5idAxYp-Asj__lVAA     2021-01-01  
3  UCT3CumbFIJiW33uq0UI3zlg     2021-01-01  
4  UCdWgRSfttvDucq4ApcCg5Mw     2021-01-01

**Question 01. 인기동영상 제작횟수가 많은 채널 상위 10개명을 출력하라 (날짜기준, 중복포함)**

In [4]:
df.channelId.value_counts().head(10)

UCszFjh7CEfwDb7UUGb4RzCQ    372
UClzB2iZ5jPoTNz0S-QU6Wiw    318
UCtm_QoN2SIxwCE-59shX7Qg    318
UCaKod3X1Tn4c7Ci0iUKcvzQ    313
UCiwQRG2sCcfjKkgxMEdJGPg    293
UChbE5OZQ6dRHECsX0tEPEZQ    283
UCLkAepWjdylmXSltofFvsYQ    275
UCQ2O-iftmnlfrBuNsUUTofQ    274
UCtfriFvIe5gIe8wRWb8UTMg    258
UCRuSxVu4iqTK5kCh90ntAgA    255
Name: channelId, dtype: int64

In [5]:
df.channelId.value_counts().head(10).index

Index(['UCszFjh7CEfwDb7UUGb4RzCQ', 'UClzB2iZ5jPoTNz0S-QU6Wiw',
       'UCtm_QoN2SIxwCE-59shX7Qg', 'UCaKod3X1Tn4c7Ci0iUKcvzQ',
       'UCiwQRG2sCcfjKkgxMEdJGPg', 'UChbE5OZQ6dRHECsX0tEPEZQ',
       'UCLkAepWjdylmXSltofFvsYQ', 'UCQ2O-iftmnlfrBuNsUUTofQ',
       'UCtfriFvIe5gIe8wRWb8UTMg', 'UCRuSxVu4iqTK5kCh90ntAgA'],
      dtype='object')

In [6]:
Ans = list(df.loc[df.channelId.isin(df.channelId.value_counts().head(10).index)].channelTitle.unique())
Ans

['장삐쭈',
 '총몇명',
 '파뿌리',
 '짤툰',
 '런닝맨 - 스브스 공식 채널',
 '엠뚜루마뚜루 : MBC 공식 종합 채널',
 'SPOTV',
 '채널 십오야',
 '이과장',
 'BANGTANTV']

**Question 02. dislikes 수가 like 수보다 높은 동영상을 제작한 채널을 모두 출력하라**

In [7]:
list(df[df['dislikes'] > df['likes']].channelTitle.unique())

['핫도그TV',
 'ASMR 애정TV',
 '하얀트리HayanTree',
 '양팡 YangPang',
 '철구형 (CHULTUBE)',
 '왜냐맨하우스',
 '(MUTUBE)와꾸대장봉준',
 '오메킴TV',
 '육지담',
 'MapleStory_KR',
 'ROAD FIGHTING CHAMPIONSHIP',
 '사나이 김기훈',
 '나혼자산다 STUDIO',
 'Gen.G esports']

**Quesiton 03. channelId의 경우 고유값이므로 이를 통해 채널명을 한 번이라도 바꾼 채널의 갯수를 구하여라**

In [8]:
df[['channelTitle', 'channelId']].drop_duplicates()

channelTitle                 channelId
0                          장삐쭈  UChbE5OZQ6dRHECsX0tEPEZQ
1      RAIN's Official Channel  UCxXgIeE5hxWxHG6dz9Scg2w
2                  서울시 · Seoul  UCZUPZW5idAxYp-Asj__lVAA
3                 고기남자 MeatMan  UCT3CumbFIJiW33uq0UI3zlg
4                        스브스밥집  UCdWgRSfttvDucq4ApcCg5Mw
...                        ...                       ...
60023                      멸치형  UCsiuN8Jco3LbbuTyGuqRgKg
60028          키키TV, 전지적 예능 시점  UCRnEWJKVc5FpC-9Z3CBt2ZQ
60035                    기용이tv  UCkMN06eC97XL1qR1bExdwWg
60036       INSSAOPPA G [인싸오빠]  UCp9G3sWsE42Qo9jgnB1ta2Q
60068              한국닌텐도 공식 채널  UCRCK5FCJtomQT3b88jXI_DA

[1843 rows x 2 columns]

In [9]:
df[['channelTitle', 'channelId']].drop_duplicates().channelId.value_counts()

UCtQYhzJWFai1y30UIA3uwYQ    3
UC98TOxKQk4aLcx0EjIK0LkQ    3
UC4nEQI74rZkFKlUSwi0ur4A    2
UCn64fSqF2l4rkkjtQUpku2A    2
UCjO3ji06DxmTM1RrzLVwmkA    2
                           ..
UCuZu8NrpBG4WPXRi-hPBl-A    1
UCDrAR1OWC2MD4s0JLetN0MA    1
UCx2MVBsdTJ_sv0LiUpSQU2Q    1
UCu2R3TCLSxYV8pRyBGNOh2g    1
UCRCK5FCJtomQT3b88jXI_DA    1
Name: channelId, Length: 1770, dtype: int64

In [10]:
change = df[['channelTitle', 'channelId']].drop_duplicates().channelId.value_counts()
target = change[change>1]
print(len(target))

71


**Question 04. 일요일에 인기있었던 영상들중 가장많은 영상 종류(categoryId)는 무엇인가?**

In [11]:
df.head()

title             channelTitle  \
0                                          [신병] 물자창고                      장삐쭈   
1   RAIN(비) - 나로 바꾸자 Switch to me (duet with JYP) MV  RAIN's Official Channel   
2       2020년 제야의 종 온라인 타종행사 | 보신각 현장 행사는 진행하지 않습니다.              서울시 · Seoul   
3                                       고기남자의 칠면조 파티             고기남자 MeatMan   
4  골목 3mc를 분노하게 만든 마음고생이 심했을 공릉 백반집 사장님의 푸념?! [예능...                    스브스밥집   

   categoryId  view_count  likes  dislikes  comment_count  \
0          23     1893473  38249       730           8595   
1          10     2600864      0         0          20129   
2          29      347049   3564       120            178   
3          26      528458  15372       280           3470   
4          24      494904   3918       111           3142   

                  channelId trending_date2  
0  UChbE5OZQ6dRHECsX0tEPEZQ     2021-01-01  
1  UCxXgIeE5hxWxHG6dz9Scg2w     2021-01-01  
2  UCZUPZW5idAxYp-Asj__lVAA     2021-01-01  
3  UCT3CumbFIJiW33uq0UI3zlg     2021-01-01  
4  UCdWgRSfttvDucq4ApcCg5Mw     2021-01-01

In [12]:
df['trending_date2'] = pd.to_datetime(df['trending_date2'])
df[df['trending_date2'].dt.day_name()=='Sunday']['categoryId'].value_counts().index[0]

24

**Question 05. 각 요일별 인기 영상들의 categoryId는 각각 몇개 씩인지 하나의 데이터 프레임으로 표현하라**

In [13]:
target = df.groupby([df['trending_date2'].dt.day_name(), 'categoryId'], as_index=False).size()
target

trending_date2  categoryId  size
0           Friday           1   243
1           Friday           2   120
2           Friday          10   833
3           Friday          15   187
4           Friday          17   633
..             ...         ...   ...
100      Wednesday          25   468
101      Wednesday          26   385
102      Wednesday          27   212
103      Wednesday          28   165
104      Wednesday          29    12

[105 rows x 3 columns]

In [14]:
Ans = target.pivot(index='categoryId', columns='trending_date2')
Ans

size                                                  
trending_date2 Friday Monday Saturday Sunday Thursday Tuesday Wednesday
categoryId                                                             
1                 243    263      255    274      246     257       234
2                 120    105      119     99      128     119       129
10                833    837      776    830      890     894       917
15                187    215      198    217      207     208       207
17                633    668      592    636      682     708       706
19                 90     92       87     91       92      89        85
20                283    298      296    289      282     285       291
22               1288   1373     1289   1337     1341    1375      1333
23                568    594      570    556      560     569       566
24               2976   3148     3066   3096     2954    3084      3090
25                444    453      422    437      470     452       468
26                369    378      364    363      375     394       385
27                183    205      183    199      194     194       212
28                171    160      173    167      166     161       165
29                 12     10       10      9       13      11        12

**Question 06.  viewcount 대비 댓글 수가 가장 높은 영상을 확인하라 (view_count값이 0인 경우는 제외한다)**

In [15]:
df.head()

title             channelTitle  \
0                                          [신병] 물자창고                      장삐쭈   
1   RAIN(비) - 나로 바꾸자 Switch to me (duet with JYP) MV  RAIN's Official Channel   
2       2020년 제야의 종 온라인 타종행사 | 보신각 현장 행사는 진행하지 않습니다.              서울시 · Seoul   
3                                       고기남자의 칠면조 파티             고기남자 MeatMan   
4  골목 3mc를 분노하게 만든 마음고생이 심했을 공릉 백반집 사장님의 푸념?! [예능...                    스브스밥집   

   categoryId  view_count  likes  dislikes  comment_count  \
0          23     1893473  38249       730           8595   
1          10     2600864      0         0          20129   
2          29      347049   3564       120            178   
3          26      528458  15372       280           3470   
4          24      494904   3918       111           3142   

                  channelId trending_date2  
0  UChbE5OZQ6dRHECsX0tEPEZQ     2021-01-01  
1  UCxXgIeE5hxWxHG6dz9Scg2w     2021-01-01  
2  UCZUPZW5idAxYp-Asj__lVAA     2021-01-01  
3  UCT3CumbFIJiW33uq0UI3zlg     2021-01-01  
4  UCdWgRSfttvDucq4ApcCg5Mw     2021-01-01

In [16]:
target = df.loc[df.view_count!=0]
targetCopy = target.copy()
targetCopy['ratio'] = (target['comment_count'] / target['view_count']).dropna()
Ans = targetCopy.sort_values(by='ratio', ascending=False).iloc[0].title
Ans

'60분 동안 댓글이 달리지 않으면, 영상이 삭제됩니다. (챌린지)'

**Question 07. viewcount 대비 댓글 수가 가장 낮은 영상을 확인하라 (view_counts, ratio값이 0인 경우는 제외한다)**

In [17]:
ratio = (target['comment_count'] / target['view_count']).dropna().sort_values()
ratio[ratio!=0].index[0]
Ans = df.iloc[ratio[ratio!=0].index[0]].title
Ans

'Join the BTS #PermissiontoDance Challenge only on YouTube #Shorts'

In [18]:
ratio

50071    0.000000
25225    0.000000
52155    0.000000
54884    0.000000
10772    0.000000
           ...   
51494    0.150621
52786    0.150693
51264    0.153998
51047    0.159655
50830    0.161516
Length: 60386, dtype: float64

In [19]:
ratio[ratio!=0].index[0]

46005

In [20]:
df.iloc[ratio[ratio!=0].index[0]]

title             Join the BTS #PermissiontoDance Challenge only...
channelTitle                                                YouTube
categoryId                                                       27
view_count                                                 46458563
likes                                                             0
dislikes                                                          0
comment_count                                                     1
channelId                                  UCBR8-60-B28hp2BmDPdntcQ
trending_date2                                  2021-08-09 00:00:00
Name: 46005, dtype: object

**Question 08. like 대비 dislike의 수가 가장 적은 영상은 무엇인가? (like, dislike 값이 0인경우는 제외한다)**

In [21]:
target = df.loc[(df.likes!=0) & (df.dislikes!=0)]
num = (target['dislikes'] / target['likes']).sort_values().index[0]
Ans = df.iloc[num].title
Ans

'[줌터뷰] *최초공개* 사부작즈🐰🐶의 비공식 이름은 아이라인즈? 꿀조합 티키타카 가득한 NCT 127 도영&정우의 줌터뷰'

In [22]:
target

title  \
0                                              [신병] 물자창고   
2           2020년 제야의 종 온라인 타종행사 | 보신각 현장 행사는 진행하지 않습니다.   
3                                           고기남자의 칠면조 파티   
4      골목 3mc를 분노하게 만든 마음고생이 심했을 공릉 백반집 사장님의 푸념?! [예능...   
5      (EN/CN/ID) 월드스타 한 분 더 모셨습니다. 비x박진영 & 싸이, 비주얼 가...   
...                                                  ...   
60394                       *풀코스로 먹습니다* 🍗치킨 + 🍕피자 + 🥧디저트   
60395  [놀면 뭐하니? 예고] 우린 깐부잖아🦑 놀뭐에 등장한 월드클래스 초대손님!! MBC...   
60396                   먹방 전쟁 : 사랑스러War (feat. 401 정육식당)   
60397  ＂너.. 광수지?＂ 협상에 거짓말부터 던지는 광수 닮은꼴 김연경 선수 [런닝맨|21...   
60398          ENHYPEN (엔하이픈) 'Tamed-Dashed' Official MV   

                channelTitle  categoryId  view_count    likes  dislikes  \
0                        장삐쭈          23     1893473    38249       730   
2                서울시 · Seoul          29      347049     3564       120   
3               고기남자 MeatMan          26      528458    15372       280   
4                      스브스밥집          24      494904     3918       111   
5      시즌비시즌 Season B Season          24     1274464    26660       368   
...                      ...         ...         ...      ...       ...   
60394                     공룡          22      538190    16671       209   
60395                놀면 뭐하니?          24     1607562    14615       258   
60396      김종국 GYM JONG KOOK          17     1125911    26490       263   
60397               스브스 예능맛집          24     1741773    14994       304   
60398            HYBE LABELS          10    13753485  1784532      9737   

       comment_count                 channelId trending_date2  
0               8595  UChbE5OZQ6dRHECsX0tEPEZQ     2021-01-01  
2                178  UCZUPZW5idAxYp-Asj__lVAA     2021-01-01  
3               3470  UCT3CumbFIJiW33uq0UI3zlg     2021-01-01  
4               3142  UCdWgRSfttvDucq4ApcCg5Mw     2021-01-01  
5               2701  UCZf4ZESHAIuRtZ-eoMSL97A     2021-01-01  
...              ...                       ...            ...  
60394           3117  UCWf4QKgnTyvmehcRqw-JldQ     2021-10-19  
60395           1141  UCx6jsZ02B4K3SECUrkgPyzg     2021-10-19  
60396           1915  UCoe-0EVDJnjlSoPK8ygcGwQ     2021-10-19  
60397            802  UCnx4Fi4cmkLGDiFfZ311wWw     2021-10-19  
60398         195461  UC3IZKseVpdzPSBaWxBxundA     2021-10-19  

[59693 rows x 9 columns]

In [23]:
(target['dislikes'] / target['likes']).sort_values()

11477     0.000523
12168     0.000598
11969     0.000602
11740     0.000620
12386     0.000626
           ...    
26887    12.657785
27108    12.807119
27355    12.936469
27578    13.009868
27789    13.082719
Length: 59693, dtype: float64

In [24]:
(target['dislikes'] / target['likes']).sort_values().index

Int64Index([11477, 12168, 11969, 11740, 12386, 51744, 51952, 51526, 52172,
            51086,
            ...
            27773, 27991, 26229, 26447, 26669, 26887, 27108, 27355, 27578,
            27789],
           dtype='int64', length=59693)

In [25]:
(target['dislikes'] / target['likes']).sort_values().index[0]

11477

**Quesiton 09. 가장 많은 트렌드 영상을 제작한 채널의 이름은 무엇인가? (날짜기준, 중복포함)**

In [26]:
df.loc[df.channelId==df.channelId.value_counts().index[0]].channelTitle.unique()[0]

'짤툰'

**Question 10. 20회(20일)이상 인기동영상 리스트에 포함된 동영상의 숫자는?**

In [27]:
(df[['title', 'channelId']].value_counts()>=20).sum()

40

# 유튜브 공범 콘텐츠 동영상 데이터

In [28]:
channel = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/channelInfo.csv')
video = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/videoInfo.csv')

display(channel.head())
display(video.head())

channelid   subcnt    viewcnt  videocnt  \
0  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  410238653       736   
1  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       736   
2  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       735   
3  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       737   
4  UCkQCwnkQfgSuPTTnw_Y7v7w  1320000  412531322       737   

                    ct channelname  
0  2021-09-30 03:01:03         꽈뚜룹  
1  2021-09-30 09:01:03         꽈뚜룹  
2  2021-09-30 15:01:03         꽈뚜룹  
3  2021-09-30 21:01:03         꽈뚜룹  
4  2021-10-01 03:01:04         꽈뚜룹

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
0  c5JQp6xafqc  1667010    30474         706            0   6587   
1  c5JQp6xafqc  1669089    30495         707            0   6589   
2  c5JQp6xafqc  1674759    30522         711            0   6596   
3  c5JQp6xafqc  1677026    30555         712            0   6604   
4  c5JQp6xafqc  1681824    30585         713            0   6600   

                    ct videoname  
0  2021-10-10 15:20:03    공범 EP1  
1  2021-10-10 15:30:03    공범 EP1  
2  2021-10-10 15:40:02    공범 EP1  
3  2021-10-10 15:50:03    공범 EP1  
4  2021-10-10 16:00:03    공범 EP1

**Question 11. 각 데이터의 ‘ct’컬럼을 시간으로 인식할수 있게 datatype을 변경하고 video 데이터의 videoname의 각 value 마다 몇 개의 데이터씩 가지고 있는지 확인하라**

In [29]:
channel['ct'] = pd.to_datetime(channel['ct'])
video['ct'] = pd.to_datetime(video['ct'])

In [30]:
video['videoname'].value_counts()

 공범 EP1    3492
 공범 EP2    3204
 공범 EP3    2568
 공범 EP4    2280
 공범 EP5    1562
 공범 EP6    1274
 공범 EP7     555
 공범 EP8     266
Name: videoname, dtype: int64

**Question 12. 수집된 각 video의 가장 최신화 된 날짜의 viewcount값을 출력하라**

In [31]:
video.sort_values(['videoname', 'ct']).drop_duplicates('videoname', keep='last')[['viewcnt', 'videoname', 'ct']].reset_index(drop=True)

viewcnt videoname                  ct
0  3180532    공범 EP1 2021-11-01 15:30:03
1  2199328    공범 EP2 2021-11-01 15:30:03
2  1671294    공범 EP3 2021-11-01 15:30:03
3  1818493    공범 EP4 2021-11-01 15:30:03
4  1503435    공범 EP5 2021-11-01 15:30:04
5  1750222    공범 EP6 2021-11-01 15:30:04
6  1630200    공범 EP7 2021-11-01 15:30:05
7  1289088    공범 EP8 2021-11-01 15:30:05

In [32]:
video.sort_values(['videoname', 'ct'])

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
10     c5JQp6xafqc     6396      680           3            0    226   
11     c5JQp6xafqc    13087     1113           7            0    273   
12     c5JQp6xafqc    19991     1446          12            0    363   
13     c5JQp6xafqc    27451     1859          19            0    479   
14     c5JQp6xafqc    36559     2306          23            0    602   
...            ...      ...      ...         ...          ...    ...   
15196  yZt-h-KcmUE  1285474    25582        3615            0  31603   
15197  yZt-h-KcmUE  1286376    25592        3617            0  31627   
15198  yZt-h-KcmUE  1287172    25597        3619            0  31643   
15199  yZt-h-KcmUE  1288138    25606        3619            0  31648   
15200  yZt-h-KcmUE  1289088    25616        3621            0  31663   

                       ct videoname  
10    2021-10-07 19:10:03    공범 EP1  
11    2021-10-07 19:20:02    공범 EP1  
12    2021-10-07 19:30:03    공범 EP1  
13    2021-10-07 19:40:03    공범 EP1  
14    2021-10-07 19:50:03    공범 EP1  
...                   ...       ...  
15196 2021-11-01 14:50:06    공범 EP8  
15197 2021-11-01 15:00:05    공범 EP8  
15198 2021-11-01 15:10:05    공범 EP8  
15199 2021-11-01 15:20:04    공범 EP8  
15200 2021-11-01 15:30:05    공범 EP8  

[15201 rows x 8 columns]

In [33]:
video.sort_values(['videoname', 'ct']).drop_duplicates('videoname', keep='last')

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
3491   c5JQp6xafqc  3180532    41043        1113            0   7674   
6695   23QQ1ru9YQg  2199328    25566         763            0   9171   
9263   GygoAk1hDU0  1671294    22719         508            0  12499   
11543  bK0_3Vbfvs8  1818493    22703         408            0  16949   
13105  hv7FBjskAu0  1503435    19926         421            0  12434   
14934  AxGKHmXyuAE  1750222    24494         389            0  13025   
13660  QUjQbgj5IXM  1630200    29335         400            0  18247   
15200  yZt-h-KcmUE  1289088    25616        3621            0  31663   

                       ct videoname  
3491  2021-11-01 15:30:03    공범 EP1  
6695  2021-11-01 15:30:03    공범 EP2  
9263  2021-11-01 15:30:03    공범 EP3  
11543 2021-11-01 15:30:03    공범 EP4  
13105 2021-11-01 15:30:04    공범 EP5  
14934 2021-11-01 15:30:04    공범 EP6  
13660 2021-11-01 15:30:05    공범 EP7  
15200 2021-11-01 15:30:05    공범 EP8

In [34]:
video.sort_values(['videoname', 'ct']).drop_duplicates('videoname', keep='last')[['viewcnt', 'videoname', 'ct']]

viewcnt videoname                  ct
3491   3180532    공범 EP1 2021-11-01 15:30:03
6695   2199328    공범 EP2 2021-11-01 15:30:03
9263   1671294    공범 EP3 2021-11-01 15:30:03
11543  1818493    공범 EP4 2021-11-01 15:30:03
13105  1503435    공범 EP5 2021-11-01 15:30:04
14934  1750222    공범 EP6 2021-11-01 15:30:04
13660  1630200    공범 EP7 2021-11-01 15:30:05
15200  1289088    공범 EP8 2021-11-01 15:30:05

In [35]:
video.sort_values(['videoname', 'ct']).drop_duplicates('videoname', keep='last')[['viewcnt', 'videoname', 'ct']].reset_index(drop=True)

viewcnt videoname                  ct
0  3180532    공범 EP1 2021-11-01 15:30:03
1  2199328    공범 EP2 2021-11-01 15:30:03
2  1671294    공범 EP3 2021-11-01 15:30:03
3  1818493    공범 EP4 2021-11-01 15:30:03
4  1503435    공범 EP5 2021-11-01 15:30:04
5  1750222    공범 EP6 2021-11-01 15:30:04
6  1630200    공범 EP7 2021-11-01 15:30:05
7  1289088    공범 EP8 2021-11-01 15:30:05

**Question 13. Channel 데이터중 2021-10-03일 이후 각 채널의 처음 기록 됐던 구독자 수(subcnt)를 출력하라**

In [36]:
channel[channel['ct']>'2021-10-03'].sort_values(['ct', 'channelname']).drop_duplicates('channelname')[['channelname', 'subcnt']].reset_index(drop=True)

channelname   subcnt
0            논리왕 전기   922000
1             야전삽짱재   257000
2        츄정ChuJeong   322000
3        김농밀의 농밀한 삶     7520
4       릴펄 Lilpearl    10100
5   와글와글 WagleWagle    55000
6               조나단   215000
7           형사!탐정되다    14900
8     Balming Tiger    54300
9     곽토리 kwak tori   471000
10              꽈뚜룹  1330000

**Question 14. 각 채널의 2021-10-03 03:00:00 ~ 2021-11-01 15:00:00 까지 구독자수 (subcnt) 의 증가량을 구하여라**

In [37]:
df.head()

title             channelTitle  \
0                                          [신병] 물자창고                      장삐쭈   
1   RAIN(비) - 나로 바꾸자 Switch to me (duet with JYP) MV  RAIN's Official Channel   
2       2020년 제야의 종 온라인 타종행사 | 보신각 현장 행사는 진행하지 않습니다.              서울시 · Seoul   
3                                       고기남자의 칠면조 파티             고기남자 MeatMan   
4  골목 3mc를 분노하게 만든 마음고생이 심했을 공릉 백반집 사장님의 푸념?! [예능...                    스브스밥집   

   categoryId  view_count  likes  dislikes  comment_count  \
0          23     1893473  38249       730           8595   
1          10     2600864      0         0          20129   
2          29      347049   3564       120            178   
3          26      528458  15372       280           3470   
4          24      494904   3918       111           3142   

                  channelId trending_date2  
0  UChbE5OZQ6dRHECsX0tEPEZQ     2021-01-01  
1  UCxXgIeE5hxWxHG6dz9Scg2w     2021-01-01  
2  UCZUPZW5idAxYp-Asj__lVAA     2021-01-01  
3  UCT3CumbFIJiW33uq0UI3zlg     2021-01-01  
4  UCdWgRSfttvDucq4ApcCg5Mw     2021-01-01

In [38]:
end = channel.loc[channel.ct.dt.strftime('%Y-%m-%d %H') == '2021-11-01 15']
start = channel.loc[channel.ct.dt.strftime('%Y-%m-%d %H') == '2021-10-03 03']

In [39]:
end_df = end[['channelname', 'subcnt']].reset_index(drop=True)
start_df = start[['channelname', 'subcnt']].reset_index(drop=True)

end_df.columns = ['channelname', 'end_sub']
start_df.columns = ['channelname', 'start_sub']

tt = pd.merge(start_df, end_df)
tt['del'] = tt['end_sub'] - tt['start_sub']
result = tt[['channelname', 'del']]
display(result)

channelname    del
0               꽈뚜룹  70000
1             야전삽짱재  11000
2        츄정ChuJeong   1000
3            논리왕 전기 -11000
4   와글와글 WagleWagle      0
5               조나단  12000
6           형사!탐정되다  10300
7        김농밀의 농밀한 삶   1540
8       릴펄 Lilpearl  11000
9     곽토리 kwak tori  -2000
10    Balming Tiger   2500

**Question 15. 공범 EP1의 비디오정보 데이터 중 수집 간격이 5분 이하, 20분 이상인 데이터 구간(해당 시점 전,후) 의 시각을 모두 출력하라**

In [40]:
ep1 = video.loc[video.videoname.str.contains('EP1')].sort_values(by='ct').reset_index(drop=True)
ep1

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
0     c5JQp6xafqc     6396      680           3            0    226   
1     c5JQp6xafqc    13087     1113           7            0    273   
2     c5JQp6xafqc    19991     1446          12            0    363   
3     c5JQp6xafqc    27451     1859          19            0    479   
4     c5JQp6xafqc    36559     2306          23            0    602   
...           ...      ...      ...         ...          ...    ...   
3487  c5JQp6xafqc  3179698    41040        1113            0   7673   
3488  c5JQp6xafqc  3179971    41042        1113            0   7673   
3489  c5JQp6xafqc  3179971    41042        1113            0   7673   
3490  c5JQp6xafqc  3180332    41043        1113            0   7674   
3491  c5JQp6xafqc  3180532    41043        1113            0   7674   

                      ct videoname  
0    2021-10-07 19:10:03    공범 EP1  
1    2021-10-07 19:20:02    공범 EP1  
2    2021-10-07 19:30:03    공범 EP1  
3    2021-10-07 19:40:03    공범 EP1  
4    2021-10-07 19:50:03    공범 EP1  
...                  ...       ...  
3487 2021-11-01 14:50:04    공범 EP1  
3488 2021-11-01 15:00:03    공범 EP1  
3489 2021-11-01 15:10:03    공범 EP1  
3490 2021-11-01 15:20:02    공범 EP1  
3491 2021-11-01 15:30:03    공범 EP1  

[3492 rows x 8 columns]

In [41]:
ep1[(ep1.ct.diff(1) >= datetime.timedelta(minutes=20)) |
    (ep1.ct.diff(1) <= datetime.timedelta(minutes=5))]

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
721   c5JQp6xafqc  2228250    34559         849            0   7191   
722   c5JQp6xafqc  2228250    34559         849            0   7191   
1636  c5JQp6xafqc  2707933    37901         988            0   7410   

                      ct videoname  
721  2021-10-13 09:41:37    공범 EP1  
722  2021-10-13 09:41:37    공범 EP1  
1636 2021-10-19 18:20:03    공범 EP1

In [42]:
ep1[ep1.index.isin([720, 721, 722, 723, 1635, 1636, 1637])]

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
720   c5JQp6xafqc  2153676    34104         830            0   7120   
721   c5JQp6xafqc  2228250    34559         849            0   7191   
722   c5JQp6xafqc  2228250    34559         849            0   7191   
723   c5JQp6xafqc  2228669    34566         849            0   7194   
1635  c5JQp6xafqc  2706691    37893         987            0   7410   
1636  c5JQp6xafqc  2707933    37901         988            0   7410   
1637  c5JQp6xafqc  2708448    37903         988            0   7410   

                      ct videoname  
720  2021-10-12 19:20:03    공범 EP1  
721  2021-10-13 09:41:37    공범 EP1  
722  2021-10-13 09:41:37    공범 EP1  
723  2021-10-13 09:50:03    공범 EP1  
1635 2021-10-19 17:50:02    공범 EP1  
1636 2021-10-19 18:20:03    공범 EP1  
1637 2021-10-19 18:30:03    공범 EP1

**Question 16. 각 에피소드의 시작날짜(년-월-일)를 에피소드 이름과 묶어 데이터프레임으로 만들고 출력하라**

In [43]:
start_date = video.sort_values(['ct', 'videoname']).drop_duplicates('videoname')[['ct', 'videoname']]
start_date['date'] = start_date.ct.dt.date
answer = start_date[['date', 'videoname']]
display(answer)

date videoname
10     2021-10-07    공범 EP1
3496   2021-10-09    공범 EP2
6696   2021-10-14    공범 EP3
9264   2021-10-16    공범 EP4
11544  2021-10-21    공범 EP5
13661  2021-10-23    공범 EP6
13106  2021-10-28    공범 EP7
14935  2021-10-30    공범 EP8

**Quesion 17. 공개된 날의 21시의 viewcnt, ct, videoname 으로 구성된 데이터 프레임을 viewcnt를 내림차순으로 정렬하여 출력하라**

In [44]:
video['time'] = video.ct.dt.hour

answer = video.loc[video['time']==21]\
        .sort_values(['ct', 'videoname'])\
        .drop_duplicates('videoname')\
        .sort_values('viewcnt', ascending=False)[['ct', 'videoname', 'viewcnt']]\
        .reset_index(drop=True)

In [45]:
answer

ct videoname  viewcnt
0 2021-10-30 21:00:08    공범 EP8   264029
1 2021-10-28 21:00:04    공범 EP7   252032
2 2021-10-16 21:00:04    공범 EP4   217674
3 2021-10-23 21:00:06    공범 EP6   213899
4 2021-10-21 21:00:04    공범 EP5   201179
5 2021-10-09 21:00:03    공범 EP2   148144
6 2021-10-14 21:00:04    공범 EP3   147183
7 2021-10-07 21:00:03    공범 EP1   117340

**Question 18. video 정보의 가장 최근 데이터들에서 각 에피소드의 싫어요/좋아요 비율을 ratio 컬럼으로 만들고 videoname, ratio로 구성된 데이터 프레임을 ratio를 오름차순으로 정렬하라**

In [46]:
target = video.sort_values('ct').drop_duplicates('videoname', keep='last')[['ct', 'videoname', 'dislikecnt', 'likecnt']]
target

ct videoname  dislikecnt  likecnt
11543 2021-11-01 15:30:03    공범 EP4         408    22703
3491  2021-11-01 15:30:03    공범 EP1        1113    41043
9263  2021-11-01 15:30:03    공범 EP3         508    22719
6695  2021-11-01 15:30:03    공범 EP2         763    25566
13105 2021-11-01 15:30:04    공범 EP5         421    19926
14934 2021-11-01 15:30:04    공범 EP6         389    24494
13660 2021-11-01 15:30:05    공범 EP7         400    29335
15200 2021-11-01 15:30:05    공범 EP8        3621    25616

In [47]:
target['ratio'] = target['dislikecnt'] / target['likecnt']
target.sort_values(by='ratio')[['videoname', 'ratio']].reset_index(drop=True)

videoname     ratio
0    공범 EP7  0.013636
1    공범 EP6  0.015881
2    공범 EP4  0.017971
3    공범 EP5  0.021128
4    공범 EP3  0.022360
5    공범 EP1  0.027118
6    공범 EP2  0.029844
7    공범 EP8  0.141357

**Question 19. 2021-11-01 00:00:00 ~ 15:00:00까지 각 에피소드별 viewcnt의 증가량을 데이터 프레임으로 만드시오**

In [48]:
start = pd.to_datetime('2021-11-01 00:00:00')
end = pd.to_datetime('2021-11-01 15:00:00')

target = video.loc[(video.ct >= start) & (video.ct <= end)].reset_index(drop=True)
target

def check(x):
    result = max(x) - min(x)
    return result

answer = target[['videoname', 'viewcnt']].groupby('videoname').agg(check)
answer

viewcnt
videoname         
 공범 EP1      13298
 공범 EP2      10300
 공범 EP3       9927
 공범 EP4       9824
 공범 EP5      10824
 공범 EP6      14141
 공범 EP7      26949
 공범 EP8      89147

**Question 20. video 데이터 중에서 중복되는 데이터가 존재한다. 중복되는 각 데이터의 시간대와 videoname 을 구하여라**

In [49]:
video[video.duplicated()==True][['ct', 'videoname']]

ct videoname
722  2021-10-13 09:41:37    공범 EP1
3927 2021-10-13 09:41:37    공범 EP2

In [50]:
# video[video.index.isin(set(video.index) - set(video.drop_duplicates().index))]

# 월드컵 출전선수 골 기록 데이터

In [51]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/worldcup/worldcupgoals.csv')

In [52]:
df.head()

Player  Goals                Years  Country
0  Miroslav Klose     16  2002-2006-2010-2014  Germany
1         Ronaldo     15       1998-2002-2006   Brazil
2     Gerd Muller     14            1970-1974  Germany
3   Just Fontaine     13                 1958   France
4            Pele     12  1958-1962-1966-1970   Brazil

**Question 21. 주어진 전체 기간의 각 나라별 골득점수 상위 5개 국가와 그 득점수를 데이터프레임형태로 출력하라**

In [53]:
df.groupby('Country').sum('Goals').sort_values('Goals', ascending=False).head(5)

Goals
Country         
Brazil       228
Germany      226
Argentina    135
Italy        123
France       115

**Question 22. 주어진 전체 기간동안 골 득점을 한 선수가 가장 많은 나라 상위 5개 국가와 그 선수 숫자를 데이터 프레임 형식으로 출력하라**

In [54]:
df.groupby('Country').size().sort_values(ascending=False).head(5)

Country
Brazil       81
Germany      78
Italy        60
Argentina    59
France       58
dtype: int64

**Question 23. 년도 표기가 4자리 숫자로 안된 케이스가 존재한다. 해당 건은 몇건인지 출력하라**

In [55]:
df.head()

Player  Goals                Years  Country
0  Miroslav Klose     16  2002-2006-2010-2014  Germany
1         Ronaldo     15       1998-2002-2006   Brazil
2     Gerd Muller     14            1970-1974  Germany
3   Just Fontaine     13                 1958   France
4            Pele     12  1958-1962-1966-1970   Brazil

In [56]:
df['yearLst'] = df['Years'].str.split('-')

def checkFour(x):
    for value in x:
        if len(str(value)) != 4:
            return False
    return True

df['check'] = df['yearLst'].apply(checkFour)

result = len(df[df.check==False])
result

45

**Question 24. Q3에서 발생한 예외 케이스를 제외한 데이터프레임을 df2라고 정의하고 데이터의 행의 숫자를 출력하라**

In [57]:
df2 = df[df.check==True].reset_index(drop=True)

In [58]:
df2.shape[0]

1250

**Question 24. 월드컵 출전횟수를 나타내는 ‘LenCup’ 컬럼을 추가하고 4회 출전한 선수의 숫자를 구하여라**

In [59]:
df2['LenCup'] = df2['yearLst'].map(lambda x: len(x))
len(df2[df2['LenCup']==4])

16

In [60]:
# df2['LenCup'] = df2['yearLst'].str.len()
# result = df2['LenCup'].value_counts()[4]

**Question 26. Yugoslavia 국가의 월드컵 출전 횟수가 2회인 선수들의 숫자를 구하여라**

In [61]:
len(df2[(df2['Country']=='Yugoslavia') & (df2['LenCup']==2)]['Player'].drop_duplicates())

7

**Question 27. 2002년도에 출전한 전체 선수는 몇 명인가?**

In [62]:
len(df2[df2.Years.str.contains('2002')==True]['Player'].drop_duplicates())

156

In [63]:
# len(df2[df2.Years.str.contains('2002')])

**Question 28. 이름에 ‘carlos’ 단어가 들어가는 선수의 숫자는 몇 명인가? (대, 소문자 구분 x)**

In [64]:
len(df2[df2['Player'].str.lower().str.contains('carlos')])

13

**Quesiton 29. 월드컵 출전 횟수가 1회 뿐인 선수들 중에서 가장 많은 득점을 올렸던 선수는 누구인가?**

In [65]:
df2[df2.LenCup==1].sort_values('Goals', ascending=False).Player.values[0]

'Just Fontaine'

In [66]:
# df2[df2['LenCup']==1][['Player', 'Goals']].sort_values('Goals', ascending=False)['Player'].head(1)

**Question 30. 월드컵 출전횟수가 1회 뿐인 선수들이 가장 많은 국가는 어디인가?**

In [67]:
df2[df2['LenCup']==1]['Country'].value_counts().index[0]

'Brazil'

# 서울시 따릉이 이용 정보 데이터

In [68]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/bicycle/seoul_bi.csv')

In [69]:
df.head()

대여일자  대여시간  대여소번호 대여구분코드 성별 연령대코드  이용건수    운동량   탄소량     이동거리  사용시간
0  2021-06-01     0   3541    정기권  F  ~10대     1   0.00  0.00     0.00     8
1  2021-06-01     0    765    정기권  F  ~10대     1  27.21  0.35  1526.81    19
2  2021-06-01     0   2637    정기권  F  ~10대     1  41.40  0.37  1608.56    18
3  2021-06-01     0   2919    정기권  F  ~10대     1   0.00  0.00     0.00    75
4  2021-06-01     0    549    정기권  F  ~10대     1  13.04  0.17   731.55     6

**Question 31. 대여일자별 데이터의 수를 데이터프레임으로 출력하고, 가장 많은 데이터가 있는 날짜를 출력하라**

In [70]:
result = df['대여일자'].value_counts().sort_index().to_frame()
result

대여일자
2021-06-01  48081
2021-06-02  56100
2021-06-03  18985
2021-06-04  56788
2021-06-05  52730
2021-06-06  47787
2021-06-07  52623
2021-06-08  55623
2021-06-09  54507
2021-06-10  36992
2021-06-11  52089
2021-06-12  48569
2021-06-13  43131
2021-06-14  54945

In [71]:
result[result['대여일자']==result['대여일자'].max()].index[0]

'2021-06-04'

In [72]:
df['대여일자'].value_counts().sort_index().to_frame()

대여일자
2021-06-01  48081
2021-06-02  56100
2021-06-03  18985
2021-06-04  56788
2021-06-05  52730
2021-06-06  47787
2021-06-07  52623
2021-06-08  55623
2021-06-09  54507
2021-06-10  36992
2021-06-11  52089
2021-06-12  48569
2021-06-13  43131
2021-06-14  54945

**Question 32. 각 일자의 요일을 표기하고 (‘Monday’ ~’Sunday’) ‘day_name’컬럼을 추가하고 이를 이용하여 각 요일별 이용 횟수의 총합을 데이터 프레임으로 출력하라**

In [73]:
df['대여일자'] = pd.to_datetime(df['대여일자'])
df['day_name'] = df['대여일자'].dt.day_name()
df['day_name'].value_counts().to_frame()

day_name
Wednesday    110607
Friday       108877
Monday       107568
Tuesday      103704
Saturday     101299
Sunday        90918
Thursday      55977

**Question 33. 각 요일별 가장 많이 이용한 대여소의 이용횟수와 대여소 번호를 데이터 프레임으로 출력하라**

In [74]:
result = df.groupby(['day_name', '대여소번호']).size().to_frame('size').sort_values(['day_name', 'size'], ascending=False).reset_index()
answer = result.drop_duplicates('day_name', keep='first').reset_index(drop=True)
display(answer)

day_name  대여소번호  size
0  Wednesday    502   282
1    Tuesday    502   267
2   Thursday   2715   137
3     Sunday    502   372
4   Saturday    502   378
5     Monday    502   242
6     Friday    502   277

In [75]:
result

day_name  대여소번호  size
0      Wednesday    502   282
1      Wednesday    207   257
2      Wednesday   2102   253
3      Wednesday   2715   253
4      Wednesday    152   223
...          ...    ...   ...
16869     Friday   3878     1
16870     Friday   4523     1
16871     Friday   9980     1
16872     Friday   9997     1
16873     Friday   9998     1

[16874 rows x 3 columns]

In [76]:
result.drop_duplicates('day_name', keep='first')

day_name  대여소번호  size
0      Wednesday    502   282
2402     Tuesday    502   267
4801    Thursday   2715   137
7197      Sunday    502   372
9611    Saturday    502   378
12031     Monday    502   242
14453     Friday    502   277

**Question 34. 나이대별 대여구분 코드의 (일일권/전체횟수) 비율을 구한 후 가장 높은 비율을 가지는 나이대를 확인하라. 일일권의 경우 일일권 과 일일권(비회원)을 모두 포함하라**

In [77]:
daily = df[df.대여구분코드.isin(['일일권', '일일권(비회원)'])].연령대코드.value_counts().sort_index()
total = df.연령대코드.value_counts().sort_index()

ratio = (daily / total).sort_values(ascending=False)

print(ratio)
print('max ratio age', ratio.index[0])

~10대    0.312525
20대     0.239913
30대     0.166702
40대     0.110585
50대     0.065414
70대~    0.048962
60대     0.042462
Name: 연령대코드, dtype: float64
max ratio age ~10대


**Question 35. 연령대별 평균 이동거리를 구하여라**

In [78]:
df[['연령대코드', '이동거리']].groupby(['연령대코드']).mean()

이동거리
연령대코드             
20대    3211.890552
30대    3341.443859
40대    3514.857416
50대    3593.668100
60대    3538.145737
70대~   3085.039641
~10대   2634.426279

**Question 36. 연령대 코드가 20대인 데이터를 추출하고, 이동거리값이 추출한 데이터의 이동거리값의 평균 이상인 데이터를 추출한다. 최종 추출된 데이터를 대여일자, 대여소 번호 순서로 내림차순 정렬 후 1행부터 200행까지의 탄소량의 평균을 소숫점 3째 자리까지 구하여라**

In [79]:
tw = df[df.연령대코드=='20대'].reset_index(drop=True)

tw_mean = tw[tw.이동거리>=tw.이동거리.mean()].reset_index(drop=True)
tw_mean['탄소량'] = tw_mean['탄소량'].astype('float')

target = tw_mean.sort_values(['대여일자', '대여소번호'], ascending=False).reset_index(drop=True).iloc[:200].탄소량

result = round(target.sum()/len(target), 3)
result

1.613

**Question 37. 6월 7일 ~10대의 “이용건수”의 중앙값은?**

In [80]:
target = df[df['대여일자']=='2021-06-07'][['연령대코드', '이용건수']]
target.이용건수.median()

1.0

In [81]:
# df[(df.연령대코드 == '~10대') & (df.대여일자 == pd.to_datetime('2021-06-07'))].이용건수.median()

**Question 38. 평일 (월~금) 출근 시간대(오전 6, 7, 8시)의 대여소별 이용 횟수를 구해서 데이터 프레임 형태로 표현한 후 각 대여시간별 이용 횟수의 상위 3개 대여소와 이용횟수를 출력하라**

In [82]:
target = df[(df.day_name.isin(['Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Monday'])) & (df.대여시간.isin([6, 7, 8]))]
result = target.groupby(['대여시간', '대여소번호']).size().to_frame('이용 횟수')
result.sort_values(['대여시간', '이용 횟수'], ascending=False).groupby('대여시간').head(3)

이용 횟수
대여시간 대여소번호       
8    2701     119
     646      115
     1152      92
7    259      104
     230       77
     726       77
6    2744      45
     1125      40
     1028      36

**Question 39. 이동거리의 평균 이상의 이동거리 값을 가지는 데이터를 추출하여 추출데이터의 이동거리의 표본표준편차 값을 구하여라**

In [83]:
df[df['이동거리']>=df['이동거리'].mean()]['이동거리'].std()

5092.139707505305

**Question 40. 남성(‘M’ or ‘m’)과 여성(‘F’ or ‘f’)의 이동거리값의 평균값을 구하여라**

In [84]:
df['sex'] = df['성별'].map(lambda x: '여' if x=='F' or x=='f' else '남')
df[['sex', '이동거리']].groupby(['sex']).mean()

이동거리
sex             
남    3209.110871
여    3468.575025

In [85]:
# df['sex'] = df['성별'].map(lambda x: '남' if x in ['M','m'] else '여')

# 전세계 행복도 지표 데이터

In [86]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/happy2/happiness.csv',encoding='utf-8')

In [87]:
df.head()

행복랭킹          나라명     점수  상대GDP  사회적지원  행복기대치  선택의 자유도    관대함  부패에 대한인식  \
0     1      Finland  7.769  1.340  1.587  0.986    0.596  0.153     0.393   
1     2      Denmark  7.600  1.383  1.573  0.996    0.592  0.252     0.410   
2     3       Norway  7.554  1.488  1.582  1.028    0.603  0.271     0.341   
3     4      Iceland  7.494  1.380  1.624  1.026    0.591  0.354     0.118   
4     5  Netherlands  7.488  1.396  1.522  0.999    0.557  0.322     0.298   

     년도  
0  2019  
1  2019  
2  2019  
3  2019  
4  2019

**Question 41. 데이터는 2018년도와 2019년도의 전세계 행복 지수를 표현한다. 각 년도의 행복랭킹 10위를 차지한 나라의 행복점수의 평균을 구하여라**

In [88]:
df[df['행복랭킹']==10]['점수'].mean()

7.259

**Question 42. 각 년도의 행복랭킹 50위 이내의 나라들의 각각의 행복점수 평균을 데이터프레임으로 표시하라**

In [89]:
df[df['행복랭킹']<=50][['년도', '점수']].groupby('년도').mean()

점수
년도           
2018  6.64678
2019  6.67002

**Question 43. 2018년도 데이터들만 추출하여 행복점수와 부패에 대한 인식에 대한 상관계수를 구하여라**

In [90]:
df[df['년도']==2018][['점수', '부패에 대한인식']].corr().iloc[0, 1]

0.40529152271510027

**Question 44. 2018년도와 2019년도의 행복랭킹이 변화하지 않은 나라명의 수를 구하여라**

In [91]:
len(df[['행복랭킹', '나라명']]) - len(df[['행복랭킹', '나라명']].drop_duplicates())

15

**Quesiton 45. 2019년도 데이터들만 추출하여 각 변수 간 상관계수를 구하고, 내림차순으로 정렬한 후 상위 5개를 데이터 프레임으로 출력하라. 컬럼명은 v1, v2, corr으로 표시하라**

In [92]:
corr2019 = df[df['년도']==2019].corr(numeric_only=True).unstack().to_frame().reset_index().dropna()

result = corr2019[corr2019[0]!=1].sort_values(0, ascending=False).drop_duplicates(0)

answer = result.head(5).reset_index(drop=True)
answer.columns = ['v1', 'v2', 'corr']

display(answer)

v1     v2      corr
0  행복기대치  상대GDP  0.835462
1  상대GDP     점수  0.793883
2  행복기대치     점수  0.779883
3  사회적지원     점수  0.777058
4  사회적지원  상대GDP  0.754906

In [93]:
df[df['년도']==2019].corr(numeric_only=True)

행복랭킹        점수     상대GDP     사회적지원     행복기대치   선택의 자유도  \
행복랭킹      1.000000 -0.989096 -0.801947 -0.767465 -0.787411 -0.546606   
점수       -0.989096  1.000000  0.793883  0.777058  0.779883  0.566742   
상대GDP    -0.801947  0.793883  1.000000  0.754906  0.835462  0.379079   
사회적지원    -0.767465  0.777058  0.754906  1.000000  0.719009  0.447333   
행복기대치    -0.787411  0.779883  0.835462  0.719009  1.000000  0.390395   
선택의 자유도  -0.546606  0.566742  0.379079  0.447333  0.390395  1.000000   
관대함      -0.047993  0.075824 -0.079662 -0.048126 -0.029511  0.269742   
부패에 대한인식 -0.351959  0.385613  0.298920  0.181899  0.295283  0.438843   
년도             NaN       NaN       NaN       NaN       NaN       NaN   

               관대함  부패에 대한인식  년도  
행복랭킹     -0.047993 -0.351959 NaN  
점수        0.075824  0.385613 NaN  
상대GDP    -0.079662  0.298920 NaN  
사회적지원    -0.048126  0.181899 NaN  
행복기대치    -0.029511  0.295283 NaN  
선택의 자유도   0.269742  0.438843 NaN  
관대함       1.000000  0.326538 NaN  
부패에 대한인식  0.326538  1.000000 NaN  
년도             NaN       NaN NaN

In [94]:
df[df['년도']==2019].corr(numeric_only=True).unstack()

행복랭킹  행복랭킹        1.000000
      점수         -0.989096
      상대GDP      -0.801947
      사회적지원      -0.767465
      행복기대치      -0.787411
                    ...   
년도    행복기대치            NaN
      선택의 자유도          NaN
      관대함              NaN
      부패에 대한인식         NaN
      년도               NaN
Length: 81, dtype: float64

**Question 46. 각 년도별 하위 행복점수의 하위 5개 국가의 평균 행복점수를 구하여라**

In [95]:
df.groupby('년도').tail(5).groupby('년도').mean(numeric_only=True)[['점수']]

점수
년도          
2018  3.1800
2019  3.1408

**Question 47. 2019년 데이터를 추출하고 해당데이터의 상대 GDP 평균 이상의 나라들과 평균 이하의 나라들의 행복점수 평균을 각각 구하고 그 차이값을 출력하라**

In [96]:
target = df[df['년도']==2019]

over = target[target.상대GDP >= target.상대GDP.mean()]['점수'].mean()
under = target[target.상대GDP <= target.상대GDP.mean()]['점수'].mean()

print(over-under)

1.390483957219251


**Question 48. 각 년도의 부패에 대한 인식을 내림차순 정렬했을 때 상위 20개 국가의 부패에 대한인식의 평균을 구하여라**

In [97]:
df.sort_values(['년도', '부패에 대한인식'], ascending=False).groupby('년도').head(20).groupby('년도').mean(numeric_only=True)[['부패에 대한인식']]

부패에 대한인식
년도            
2018    0.3267
2019    0.3201

**Question 49. 2018년도 행복랭킹 50위 이내에 포함됐다가 2019년 50위 밖으로 밀려난 국가의 숫자를 구하여라**

In [98]:
target18 = df[(df['년도']==2018) & (df['행복랭킹']<=50)]['나라명']
target19 = df[(df['년도']==2019) & (df['행복랭킹']<=50)]['나라명']

len(set(target18) - set(target19))

4

In [99]:
set(target18) - set(target19)

{'Belize', 'Kuwait', 'Malaysia', 'Thailand'}

**Question 50. 2018년, 2019년 모두 기록이 있는 나라들 중 년도별 행복점수가 가장 증가한 나라와 그 증가 수치는?**

In [100]:
count = df.나라명.value_counts()
target = count[count>=2].index

df2 = df.copy()
mul = df2[df2.나라명.isin(target).reset_index(drop=True)]
mul.loc[mul['년도']==2018, '점수'] = mul[mul.년도==2018]['점수'].values * (-1)
result = mul.groupby('나라명').sum()['점수'].sort_values().to_frame().iloc[-1]
result

점수    0.87
Name: Burundi, dtype: float64

In [101]:
mul.loc[mul['년도']==2018, '점수'] 

156   -7.632
157   -7.594
158   -7.555
159   -7.495
160   -7.487
       ...  
307   -3.355
308   -3.303
309   -3.254
310   -3.083
311   -2.905
Name: 점수, Length: 152, dtype: float64

In [102]:
mul[mul.년도==2018]['점수'].values * (-1)

array([7.632, 7.594, 7.555, 7.495, 7.487, 7.441, 7.328, 7.324, 7.314,
       7.272, 7.19 , 7.139, 7.072, 6.977, 6.965, 6.927, 6.91 , 6.886,
       6.814, 6.774, 6.711, 6.627, 6.489, 6.488, 6.476, 6.441, 6.43 ,
       6.419, 6.388, 6.382, 6.379, 6.374, 6.371, 6.343, 6.322, 6.31 ,
       6.26 , 6.192, 6.173, 6.167, 6.141, 6.123, 6.105, 6.096, 6.083,
       6.072, 6.   , 5.973, 5.952, 5.948, 5.945, 5.933, 5.915, 5.891,
       5.89 , 5.875, 5.835, 5.81 , 5.79 , 5.762, 5.752, 5.739, 5.681,
       5.663, 5.662, 5.64 , 5.636, 5.62 , 5.566, 5.524, 5.504, 5.483,
       5.483, 5.472, 5.43 , 5.41 , 5.398, 5.358, 5.358, 5.347, 5.321,
       5.302, 5.295, 5.254, 5.246, 5.201, 5.199, 5.161, 5.155, 5.131,
       5.129, 5.125, 5.103, 5.093, 5.082, 4.982, 4.975, 4.933, 4.88 ,
       4.806, 4.758, 4.743, 4.724, 4.707, 4.671, 4.657, 4.631, 4.623,
       4.592, 4.586, 4.571, 4.559, 4.5  , 4.471, 4.456, 4.447, 4.441,
       4.433, 4.424, 4.419, 4.417, 4.41 , 4.377, 4.356, 4.35 , 4.34 ,
       4.321, 4.308,

In [103]:
mul.groupby('나라명').sum()['점수'].sort_values().to_frame()

점수
나라명                
Malaysia     -0.983
Afghanistan  -0.429
South Sudan  -0.401
Turkmenistan -0.389
Somalia      -0.314
...             ...
Niger         0.462
Liberia       0.480
Guinea        0.570
Benin         0.742
Burundi       0.870

[152 rows x 1 columns]

In [104]:
df[df['나라명']=='Burundi']

행복랭킹      나라명     점수  상대GDP  사회적지원  행복기대치  선택의 자유도    관대함  부패에 대한인식    년도
144   145  Burundi  3.775  0.046  0.447  0.380    0.220  0.176     0.180  2019
311   156  Burundi  2.905  0.091  0.627  0.145    0.065  0.149     0.076  2018

# 지역구 에너지 소비량 데이터

In [105]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/consum/Tetuan%20City%20power%20consumption.csv')

In [106]:
df.head()

DateTime  Temperature  Humidity  Wind Speed  general diffuse flows  \
0  1/1/2017 0:00        6.559      73.8       0.083                  0.051   
1  1/1/2017 0:10        6.414      74.5       0.083                  0.070   
2  1/1/2017 0:20        6.313      74.5       0.080                  0.062   
3  1/1/2017 0:30        6.121      75.0       0.083                  0.091   
4  1/1/2017 0:40        5.921      75.7       0.081                  0.048   

   diffuse flows  Zone 1 Power Consumption  Zone 2  Power Consumption  \
0          0.119               34055.69620                16128.87538   
1          0.085               29814.68354                19375.07599   
2          0.100               29128.10127                19006.68693   
3          0.096               28228.86076                18361.09422   
4          0.085               27335.69620                17872.34043   

   Zone 3  Power Consumption  
0                20240.96386  
1                20131.08434  
2                19668.43373  
3                18899.27711  
4                18442.40964

**Question 51. DateTime 컬럼을 통해 각 월별로 몇 개의 데이터가 있는지 데이터 프레임으로 구하여라**

In [107]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52416 entries, 0 to 52415
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   DateTime                   52416 non-null  object 
 1   Temperature                52416 non-null  float64
 2   Humidity                   52416 non-null  float64
 3   Wind Speed                 52416 non-null  float64
 4   general diffuse flows      52416 non-null  float64
 5   diffuse flows              52416 non-null  float64
 6   Zone 1 Power Consumption   52416 non-null  float64
 7   Zone 2  Power Consumption  52416 non-null  float64
 8   Zone 3  Power Consumption  52416 non-null  float64
dtypes: float64(8), object(1)
memory usage: 3.6+ MB


In [108]:
df['DateTime'] = pd.to_datetime(df['DateTime'])

In [109]:
df['DateTime'].dt.month.value_counts().sort_index()

1     4464
2     4032
3     4464
4     4320
5     4464
6     4320
7     4464
8     4464
9     4320
10    4464
11    4320
12    4320
Name: DateTime, dtype: int64

**Question 52. 3월달의 각 시간대별 온도의 평균들 중 가장 낮은 시간대의 온도를 출력하라**

In [110]:
target = df[df['DateTime'].dt.month == 3]
result = target.groupby(target.DateTime.dt.hour)['Temperature'].mean().sort_values().values[0]
result

11.506612903225806

**Question 53. 3월달의 각 시간대별 온도의 평균들 중 가장 높은 시간대의 온도를 출력하라**

In [111]:
target.groupby(target.DateTime.dt.hour)['Temperature'].mean().sort_values().values[-1]

18.393602150537635

**Question 54. Zone 1 Power Consumption 컬럼의 value값의 크기가 Zone 2 Power Consumption 컬럼의 value값의 크기보다 큰 데이터들의 Humidity의 평균을 구하여라**

In [112]:
df[(df['Zone 1 Power Consumption']>df['Zone 2  Power Consumption'])]['Humidity'].mean()

68.23624448055052

**Question 55. 각 zone의 에너지 소비량의 상관관계를 구해서 데이터 프레임으로 표기하라**

In [113]:
df.iloc[:, -3:].corr()

Zone 1 Power Consumption  \
Zone 1 Power Consumption                   1.000000   
Zone 2  Power Consumption                  0.834519   
Zone 3  Power Consumption                  0.750733   

                           Zone 2  Power Consumption  \
Zone 1 Power Consumption                    0.834519   
Zone 2  Power Consumption                   1.000000   
Zone 3  Power Consumption                   0.570932   

                           Zone 3  Power Consumption  
Zone 1 Power Consumption                    0.750733  
Zone 2  Power Consumption                   0.570932  
Zone 3  Power Consumption                   1.000000

**Question 56. Temperature의 값이 10미만의 경우 A, 10이상 20미만의 경우 B,20이상 30미만의 경우 C, 그 외의 경우 D라고 할때 각 단계의 데이터 숫자를 구하여라**

In [114]:
def grade(x):
    if x < 10:
        return 'A'
    elif x < 20:
        return 'B'
    elif x < 30:
        return 'C'
    else:
        return 'D'
    
df['grade'] = df['Temperature'].apply(grade)
df['grade'].value_counts()

B    26993
C    21105
A     2874
D     1444
Name: grade, dtype: int64

**Question 57. 6월 데이터중 12시의 Temperature의 표준편차를 구하여라**

In [115]:
target = df[df['DateTime'].dt.month == 6]
target[target['DateTime'].dt.hour==12]['Temperature'].std()

2.049941782795103

In [116]:
# df[(df.DateTime.dt.month ==6) & (df.DateTime.dt.hour ==12)].Temperature.std()

**Question 58. 6월 데이터중 12시의 Temperature의 분산을 구하여라**

In [117]:
target[target['DateTime'].dt.hour==12]['Temperature'].var()

4.202261312849164

**Question 59. Temperature의 평균 이상의 Temperature의 값을 가지는 데이터를 Temperature를 기준으로 정렬 했을때 4번째 행의 Humidity 값은?**

In [118]:
df[df['Temperature']>=df['Temperature'].mean()].sort_values(by='Temperature')['Humidity'].values[3]

87.9

**Question 60. Temperature의 중간값 이상의 Temperature의 값을 가지는 데이터를Temperature를 기준으로 정렬 했을때 4번째 행의 Humidity 값은?**

In [119]:
df[df['Temperature']>=df['Temperature'].median()].sort_values(by='Temperature')['Humidity'].values[3]

80.3

# 포켓몬 정보 데이터

In [120]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/pok/Pokemon.csv')

In [121]:
df.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

**Question 61. Legendary 컬럼은 전설 포켓몬 유무를 나타낸다. 전설 포켓몬과 그렇지 않은 포켓몬들의 HP평균의 차이를 구하여라**

In [122]:
target = df.groupby('Legendary').mean(numeric_only=True)['HP']
result = target.values[1] -target.values[0]
print(result)

25.55614861329147


In [123]:
# df.groupby('Legendary')['HP'].mean()[1] - df.groupby('Legendary')['HP'].mean()[0]

**Question 62. Type 1은 주속성 Type 2 는 부속성을 나타낸다. 가장 많은 부속성 종류는 무엇인가?**

In [124]:
df['Type 2'].value_counts().index[0]

'Flying'

**Question 63. 가장 많은 Type 1 의 종의 평균 Attack 을 평균 Defense로 나눈값은?**

In [125]:
ta = df[df['Type 1'] == df['Type 1'].value_counts().index[0]]['Attack'].mean()
td = df[df['Type 1'] == df['Type 1'].value_counts().index[0]]['Defense'].mean()

ta / td

1.0165238678090576

In [126]:
max = df['Type 1'].value_counts().index[0]
result = df[df['Type 1']== max].Attack.mean() /df[df['Type 1']== max].Defense.mean()
print(result)

1.0165238678090576


**Question 64. 포켓몬 세대(Generation) 중 가장 많은 Legendary를 보유한 세대는 몇 세대인가?**

In [127]:
df[df.Legendary==True].Generation.value_counts().index[0]

3

In [128]:
# df[df['Legendary']==True].groupby('Generation').size().sort_values().index[-1]

**Question 65. ‘HP’, ‘Attack’, ‘Defense’, ‘Sp. Atk’, ‘Sp. Def’, ‘Speed’ 간의 상관계수중 가장 절댓값이 큰 두 변수와 그 값을 구하여라**

In [129]:
target = df[['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']].corr().unstack().reset_index().rename(columns={0: 'corr'})
target[target['corr']!=1].sort_values(by='corr').iloc[-1]

level_0     Defense
level_1     Sp. Def
corr       0.510747
Name: 16, dtype: object

**Question 66. 각 Generation의 Attack으로 오름차순 정렬시 상위 3개 데이터들(18개)의 Attack의 전체 평균을 구하여라**

In [130]:
df.sort_values(['Generation','Attack']).groupby('Generation').head(3).Attack.mean()

19.5

**Question 67. 각 Generation의 Attack으로 내림차순 정렬시 상위 5개 데이터들(30개)의 Attack의 전체 평균을 구하여라**

In [131]:
df.sort_values(['Generation', 'Attack'], ascending=False).groupby('Generation').head(5).Attack.mean()

157.23333333333332

**Question 68. 가장 흔하게 발견되는 (Type1 , Type2) 의 쌍은 무엇인가?**

In [132]:
df[['Type 1', 'Type 2']].value_counts().head(1)

Type 1  Type 2
Normal  Flying    24
dtype: int64

**Question 69. 한 번씩만 존재하는 (Type1 , Type2)의 쌍의 갯수는 몇 개인가?**

In [133]:
target = df[['Type 1','Type 2']].value_counts()
result = len(target[target==1])
print(result)

39


In [134]:
target = df[['Type 1', 'Type 2']].value_counts().to_frame()
len(target[target[0]==1])

39

**Question 70. 한 번씩만 존재하는 (Type1 , Type2)의 쌍을 각 세대(Generation)은 각각 몇 개씩 가지고 있는가?**

In [135]:
target = df[['Type 1', 'Type 2']].value_counts()
target2 = target[target==1]

lst = []
for value in target2.reset_index().values:
    t1 = value[0]
    t2 = value[1]
    
    sp = df[(df['Type 1']==t1) & (df['Type 2']==t2)]
    lst.append(sp)

result = pd.concat(lst).reset_index(drop=True).Generation.value_counts().sort_index()
result

1     1
2     4
3     5
4    13
5     7
6     9
Name: Generation, dtype: int64

# 대한민국 체력장 데이터

In [136]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/body/body.csv')

In [137]:
df.head()

측정나이 측정회원성별  신장 : cm  체중 : kg  체지방율 : %  이완기혈압(최저) : mmHg  \
0  59.0      M    175.0     70.6      19.2              91.0   
1  40.0      F    161.4     45.8      21.7              62.0   
2  27.0      M    176.6     84.8      19.2              87.0   
3  38.0      M    167.6     76.7      24.7              63.0   
4  21.0      M    165.2     66.2      21.5              83.0   

   수축기혈압(최고) : mmHg  악력D : kg  앉아윗몸앞으로굽히기 : cm  교차윗몸일으키기 : 회  제자리 멀리뛰기 : cm 등급  
0             150.0      40.6             12.2          30.0          179.0  C  
1             119.0      22.1             16.0          32.0          165.0  B  
2             153.0      45.9             13.2          61.0          216.0  B  
3             132.0      43.6             16.0          45.0          231.0  A  
4             106.0      33.5             10.6          46.0          198.0  C

**Question 71. 전체 데이터의 수축기혈압(최고) - 이완기혈압(최저)의 평균을 구하여라**

In [138]:
(df['수축기혈압(최고) : mmHg'] - df['이완기혈압(최저) : mmHg']).mean()

52.19539414750672

**Question 72. 50~59세의 신장평균을 구하여라**

In [139]:
df[(df['측정나이']>= 50) & (df['측정나이']<60)]['신장 : cm'].mean()

164.07490107405314

**Question 73. 연령대 (20~29 : 20대 …) 별 인원수를 구하여라**

In [140]:
df['연령대'] = df['측정나이'].map(lambda x: x//10*10)
df['연령대'].value_counts()

20.0    5831
30.0    2660
40.0    1801
50.0    1769
60.0    1335
Name: 연령대, dtype: int64

**Question 74. 연령대 (20~29 : 20대 …) 별 등급의 숫자를 데이터 프레임으로 표현하라**

In [141]:
df.groupby(['연령대', '등급'], as_index=False).size()

연령대 등급  size
0   20.0  A  1585
1   20.0  B  1443
2   20.0  C  1455
3   20.0  D  1348
4   30.0  A   743
5   30.0  B   697
6   30.0  C   626
7   30.0  D   594
8   40.0  A   386
9   40.0  B   428
10  40.0  C   455
11  40.0  D   532
12  50.0  A   321
13  50.0  B   410
14  50.0  C   474
15  50.0  D   564
16  60.0  A   314
17  60.0  B   371
18  60.0  C   339
19  60.0  D   311

In [142]:
# df[['연령대', '등급']].value_counts().reset_index()

**Question 75. 남성 중 A등급과 D등급의 체지방률 평균의 차이(큰 값에서 작은 값의 차)를 구하여라**

In [143]:
ma = df[(df['측정회원성별']=='M') & (df['등급']=='A')]['체지방율 : %'].mean()
md = df[(df['측정회원성별']=='M') & (df['등급']=='D')]['체지방율 : %'].mean()

abs(ma-md)

7.932086486137521

**Question 76. 여성 중 A등급과 D등급의 체중의 평균의 차이(큰 값에서 작은 값의 차)를 구하여라**

In [144]:
ma = df[(df['측정회원성별']=='F') & (df['등급']=='A')]['체중 : kg'].mean()
md = df[(df['측정회원성별']=='F') & (df['등급']=='D')]['체중 : kg'].mean()

abs(ma-md)

5.176211590296504

**Question 77. bmi는 자신의 몸무게(kg)를 키의 제곱(m)으로 나눈값이다. 데이터의 bmi 를 구한 새로운 컬럼을 만들고 남성의 bmi 평균을 구하여라**

In [145]:
df['bmi'] = df['체중 : kg'] / (df['신장 : cm']/100)**2
df[df['측정회원성별']=='M']['bmi'].mean()

24.46134409819308

**Question 78. bmi보다 체지방율이 높은 사람들의 체중평균을 구하여라**

In [146]:
df[df['체지방율 : %']>df['bmi']]['체중 : kg'].mean()

61.7408806392542

**Question 79. 남성과 여성의 악력 평균의 차이를 구하여라**

In [147]:
target = df.groupby('측정회원성별')['악력D : kg'].mean()
result = target.M - target.F
result

17.55954185047465

In [148]:
md = df[df['측정회원성별']=='M']['악력D : kg'].mean()
fm = df[df['측정회원성별']=='F']['악력D : kg'].mean()

abs(md-fm)

17.55954185047464

**Question 80. 남성과 여성의 교차윗몸일으키기 횟수의 평균의 차이를 구하여라**

In [149]:
target = df.groupby('측정회원성별')['교차윗몸일으키기 : 회'].mean()
abs(target.M - target.F)

14.243156833157627

# 기온 강수량 데이터

In [150]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/weather/weather2.csv")

In [151]:
df.head()

time  이화동강수  이화동기온  수영동강수  수영동기온
0  2020-01-01 00:00:00    0.1   -3.9    0.0    2.5
1  2020-01-01 01:00:00    0.0   -3.1    0.0    3.4
2  2020-01-01 02:00:00    0.0   -1.8    0.0    3.8
3  2020-01-01 03:00:00    0.0   -0.8    0.0    4.6
4  2020-01-01 04:00:00    0.0   -0.1    0.0    5.1

**Question 81. 여름철(6월, 7월, 8월) 이화동이 수영동보다 높은 기온을 가진 시간대는 몇개인가?**

In [152]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   time    8784 non-null   object 
 1   이화동강수   8784 non-null   float64
 2   이화동기온   8784 non-null   float64
 3   수영동강수   8784 non-null   float64
 4   수영동기온   8784 non-null   float64
dtypes: float64(4), object(1)
memory usage: 343.2+ KB


In [153]:
df.time = pd.to_datetime(df.time)
summer = df.loc[df.time.dt.month.isin([6,7,8])].reset_index(drop=True)
answer = len(summer.loc[summer['이화동기온'] > summer['수영동기온']])
print(answer)

1415


In [154]:
df['time'] = pd.to_datetime(df['time'])
target = df[(df['time'].dt.month.isin([6, 7, 8])) & (df['이화동기온']>df['수영동기온'])]
len(target.time.dt.hour)

1415

**Question 82. 이화동과 수영동의 최대 강수량의 시간대를 각각 구하여라**

In [155]:
answer1 = df.loc[df['이화동강수']==df['이화동강수'].max()].time.values
answer2 = df.loc[df['수영동강수']==df['수영동강수'].max()].time.values
print(answer1, answer2)

['2020-09-30T09:00:00.000000000'] ['2020-07-23T12:00:00.000000000']


In [156]:
# print(df[df.이화동강수 == df.이화동강수.max()]['time'], df[df.수영동강수 == df.수영동강수.max()]['time'])

# 서비스 이탈예측 데이터

In [157]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/churn/train.csv")

In [158]:
df.head()

RowNumber  CustomerId     Surname  CreditScore Geography  Gender  Age  \
0       6842    15793491  Cherkasova          714   Germany    Male   26   
1       8963    15607874       Keane          687    France    Male   38   
2       7047    15737627      Rivero          589   Germany  Female   20   
3       7503    15697844  Whitehouse          721     Spain  Female   32   
4       3439    15722404   Carpenter          445    France  Female   30   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       3  119545.48              2          1               0   
1       0  144450.58              1          0               1   
2       2  121093.29              2          1               0   
3      10       0.00              1          1               0   
4       3       0.00              2          1               1   

   EstimatedSalary  Exited  
0         65482.94       0  
1        137276.83       0  
2          3529.72       0  
3        136119.96       1  
4        127939.19       0

**Question 83. 남성 이탈(Exited)이 가장 많은 국가(Geography)는 어디이고 이탈 인원은 몇명인가?**

In [159]:
answer = df.loc[df.Gender=='Male'].groupby(['Geography']).sum().Exited.sort_values(ascending=False).head(1)
print(answer)

Geography
Germany    287
Name: Exited, dtype: int64


C:\Users\futur\AppData\Local\Temp\ipykernel_8792\3947877788.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  answer = df.loc[df.Gender=='Male'].groupby(['Geography']).sum().Exited.sort_values(ascending=False).head(1)


In [160]:
df[(df['Gender']=='Male') & (df['Exited'] == 1)]['Geography'].value_counts().head(1)

Germany    287
Name: Geography, dtype: int64

**Question 84. 카드를 소유(HasCrCard==1)하고 있으면서 활성멤버(IsActiveMember==1)인 고객들의 평균 나이를 소숫점이하 4자리까지 구하여라**

In [161]:
round(df[(df['HasCrCard']==1) & (df['IsActiveMember']==1)]['Age'].mean(), 4)

39.6102

**Question 85. Balance 값이 중간값 이상을 가지는 고객들의 CreditScore의 표준편차를 소숫점이하 3자리까지 구하여라**

In [162]:
round(df[df['Balance'] >= df['Balance'].median()]['CreditScore'].std(), 3)

97.295

# 성인 건강검진 데이터

In [163]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/smoke/train.csv")

In [164]:
df.head()

성별코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg단위)  허리둘레  시력(좌)  시력(우)  청력(좌)  청력(우)  \
0    F           55        145         55  73.0    0.7    0.5    1.0    1.0   
1    M           40        180         55  74.0    1.5    1.5    1.0    1.0   
2    F           55        150         50  72.0    1.0    0.2    1.0    1.0   
3    M           40        170         85  89.0    1.0    0.8    1.0    1.0   
4    F           40        155         45  62.0    0.5    1.0    1.0    1.0   

   수축기혈압  ...   혈색소  요단백  혈청크레아티닌  (혈청지오티)AST  (혈청지오티)ALT  감마지티피  흡연상태  \
0  129.0  ...  12.9  1.0      0.9       172.0       209.0   15.0     0   
1  102.0  ...  15.6  1.0      0.8        30.0        19.0   23.0     1   
2  116.0  ...  14.1  1.0      1.0        31.0        19.0   10.0     0   
3  124.0  ...  16.7  1.0      1.0        26.0        38.0  108.0     1   
4  127.0  ...  12.9  1.0      0.7        20.0        11.0   13.0     0   

   구강검진수검여부  치아우식증유무  치석  
0         Y      0.0   Y  
1         Y      0.0   Y  
2         Y      0.0   N  
3         Y      1.0   Y  
4         Y      0.0   N  

[5 rows x 26 columns]

In [165]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44553 entries, 0 to 44552
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   성별코드         44553 non-null  object 
 1   연령대코드(5세단위)  44553 non-null  int64  
 2   신장(5Cm단위)    44553 non-null  int64  
 3   체중(5Kg단위)    44553 non-null  int64  
 4   허리둘레         44553 non-null  float64
 5   시력(좌)        44553 non-null  float64
 6   시력(우)        44553 non-null  float64
 7   청력(좌)        44553 non-null  float64
 8   청력(우)        44553 non-null  float64
 9   수축기혈압        44553 non-null  float64
 10  이완기혈압        44553 non-null  float64
 11  식전혈당(공복혈당)   44553 non-null  float64
 12  총콜레스테롤       44553 non-null  float64
 13  트리글리세라이드     44553 non-null  float64
 14  HDL콜레스테롤     44553 non-null  float64
 15  LDL콜레스테롤     44553 non-null  float64
 16  혈색소          44553 non-null  float64
 17  요단백          44553 non-null  float64
 18  혈청크레아티닌      44553 non-null  float64
 19  (혈청지

**Question 86. 수축기혈압과 이완기 혈압기 수치의 차이를 새로운 컬럼(‘혈압차’) 으로 생성하고, 연령대 코드별 각 그룹 중 ‘혈압차’ 의 분산이 5번째로 큰 연령대 코드를 구하여라**

In [166]:
df['혈압차'] = df['수축기혈압'] - df['이완기혈압']
df.groupby('연령대코드(5세단위)').var(numeric_only=True)['혈압차'].sort_values(ascending=False).index[4]

60

In [167]:
df.groupby('연령대코드(5세단위)')['혈압차'].var().sort_values().index[-5]

60

**Question 87. 비만도를 나타내는 지표인 WHtR는 허리둘레 / 키로 표현한다. 일반적으로 0.58이상이면 비만으로 분류한다. 데이터중 WHtR 지표상 비만인 인원의 남/여 비율을 구하여라**

In [168]:
df['WHtR'] = df['허리둘레'] / df['신장(5Cm단위)']
target = df[df['WHtR']>=0.58].groupby('성별코드').size()
target.M/target.F

1.1693877551020408

In [169]:
data = df.loc[df['WHtR']>=0.58].성별코드.value_counts()
answer = data['M']/data['F']
print(answer)

1.1693877551020408


# 자동차 보험가입 예측 데이터

In [170]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/insurance/train.csv")

In [171]:
df.head()

id  Gender  Age  Driving_License  Region_Code  Previously_Insured  \
0  188957  Female   25                1         17.0                   1   
1  275631    Male   20                1         22.0                   0   
2  329036  Female   40                1         41.0                   0   
3  227288  Female   28                1         30.0                   1   
4  300441  Female   50                1         28.0                   0   

  Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel  Vintage  \
0    < 1 Year             No         38785.0                 152.0       59   
1    < 1 Year             No          2630.0                 160.0      113   
2    1-2 Year            Yes         33165.0                 124.0      188   
3    < 1 Year             No         29038.0                 152.0      233   
4    1-2 Year            Yes         31325.0                 124.0      258   

   Response  
0         0  
1         0  
2         0  
3         0  
4         0

**Question 88. Vehicle_Age 값이 2년 이상인 사람들만 필터링 하고 그중에서
Annual_Premium 값이 전체 데이터의 중간값 이상인 사람들을 찾고, 그들의 Vintage값의 평균을 구하여라**

In [172]:
df['Vehicle_Age'].unique()

array(['< 1 Year', '1-2 Year', '> 2 Years'], dtype=object)

In [173]:
df[(df['Vehicle_Age']=='> 2 Years') & (df['Annual_Premium']>=df['Annual_Premium'].median())]['Vintage'].mean()

154.43647182359118

**Question 89. vehicle_age에 따른 각 성별(gender)그룹의 Annual_Premium값의 평균을 구하여 아래 테이블과 동일하게 구현하라**

In [174]:
meandf = df.groupby(['Gender', 'Vehicle_Age'], as_index=False)['Annual_Premium'].mean()
meandf.pivot(index='Vehicle_Age', columns='Gender', values='Annual_Premium')

Gender             Female          Male
Vehicle_Age                            
1-2 Year     30762.245001  30413.088469
< 1 Year     29972.286702  30310.982212
> 2 Years    36108.366374  35303.870627

In [175]:
df.pivot_table(index='Vehicle_Age', columns='Gender', values='Annual_Premium', aggfunc='mean')

Gender             Female          Male
Vehicle_Age                            
1-2 Year     30762.245001  30413.088469
< 1 Year     29972.286702  30310.982212
> 2 Years    36108.366374  35303.870627

# 핸드폰 가격 예측데이터

In [176]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/mobile/train.csv")

In [177]:
df.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

**Question 90. price_range 의 각 value를 그룹핑하여 각 그룹의 n_cores 의 빈도가 가장높은 value와 그 빈도수를 구하여라**

In [178]:
df[['price_range', 'n_cores']].groupby(['price_range', 'n_cores']).size().sort_values(0).groupby(level=0).tail(1)

C:\Users\futur\AppData\Local\Temp\ipykernel_8792\2738451872.py:1: FutureWarning: In a future version of pandas all arguments of Series.sort_values will be keyword-only.
  df[['price_range', 'n_cores']].groupby(['price_range', 'n_cores']).size().sort_values(0).groupby(level=0).tail(1)


price_range  n_cores
0            2          69
3            5          70
2            4          73
1            1          76
dtype: int64

In [179]:
df[['price_range', 'n_cores']].groupby(['price_range', 'n_cores']).size().sort_values(0)

C:\Users\futur\AppData\Local\Temp\ipykernel_8792\2257681911.py:1: FutureWarning: In a future version of pandas all arguments of Series.sort_values will be keyword-only.
  df[['price_range', 'n_cores']].groupby(['price_range', 'n_cores']).size().sort_values(0)


price_range  n_cores
0            1          49
1            5          51
2            1          52
1            6          54
             7          55
2            3          56
             6          57
             2          57
3            6          58
             4          58
             3          59
             8          59
0            5          59
1            2          59
             8          60
0            6          61
             3          62
3            2          62
             1          65
0            7          66
2            5          66
0            8          67
             4          67
3            7          69
0            2          69
1            3          69
2            7          69
3            5          70
2            8          70
             4          73
1            4          76
             1          76
dtype: int64

**Question 91. price_range 값이 3인 그룹에서 상관관계가 2번째로 높은 두 컬럼과 그 상관계수를 구하여라**

In [180]:
cordf = df.loc[df['price_range']==3].corr().unstack().sort_values(ascending=False)
answer  = cordf.loc[cordf!=1].reset_index().iloc[1]
print(answer)

level_0          pc
level_1          fc
0          0.635166
Name: 1, dtype: object


In [181]:
target = df[df['price_range']==3].corr().unstack().to_frame()
target[target[0]!=1].sort_values(0, ascending=False).reset_index().iloc[1, :]

level_0          fc
level_1          pc
0          0.635166
Name: 1, dtype: object

# 비행 탑승 경험 만족도 데이터

In [182]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/airline/train.csv")

In [183]:
df.head()

id  Gender   Customer Type  Age   Type of Travel     Class  \
0   55882    Male  Loyal Customer   42  Business travel  Business   
1   31906  Female  Loyal Customer   31  Business travel  Business   
2   85481    Male  Loyal Customer   51  Personal Travel       Eco   
3  102134    Male  Loyal Customer   44  Business travel  Business   
4   31580    Male  Loyal Customer   33  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0              473                      1                                  1   
1             3271                      4                                  4   
2              214                      5                                  0   
3             3097                      1                                  1   
4             3286                      1                                  1   

   Ease of Online booking  ...  Inflight entertainment  On-board service  \
0                       3  ...                       5                 5   
1                       4  ...                       3                 3   
2                       5  ...                       3                 4   
3                       3  ...                       4                 4   
4                       1  ...                       4                 2   

   Leg room service  Baggage handling  Checkin service  Inflight service  \
0                 5                 5                3                 5   
1                 5                 5                3                 4   
2                 4                 4                4                 4   
3                 4                 4                3                 4   
4                 5                 2                5                 1   

   Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  \
0            3                           3                       0.0   
1            3                           1                       5.0   
2            3                           2                       0.0   
3            5                           0                       0.0   
4            4                           0                       0.0   

   satisfaction  
0     satisfied  
1     satisfied  
2     satisfied  
3     satisfied  
4     satisfied  

[5 rows x 24 columns]

**Question 92. Arrival Delay in Minutes 컬럼이 결측치인 데이터들 중 ‘neutral or dissatisfied’ 보다 ‘satisfied’의 수가 더 높은 Class는 어디 인가?**

In [184]:
answer = df.loc[df['Arrival Delay in Minutes'].isnull()].groupby(['Class', 'satisfaction'], as_index=False).size().pivot(index='Class', columns='satisfaction')
result = answer[answer['size']['neutral or dissatisfied'] < answer['size']['satisfied']]
result

size          
satisfaction neutral or dissatisfied satisfied
Class                                         
Business                          36        76

In [185]:
df.loc[df['Arrival Delay in Minutes'].isnull()].groupby(['Class', 'satisfaction'], as_index=False).size()

Class             satisfaction  size
0  Business  neutral or dissatisfied    36
1  Business                satisfied    76
2       Eco  neutral or dissatisfied   100
3       Eco                satisfied    25
4  Eco Plus  neutral or dissatisfied    15
5  Eco Plus                satisfied     4

In [186]:
df.loc[df['Arrival Delay in Minutes'].isnull()].groupby(['Class', 'satisfaction'], as_index=False).size().pivot(index='Class', columns='satisfaction')

size          
satisfaction neutral or dissatisfied satisfied
Class                                         
Business                          36        76
Eco                              100        25
Eco Plus                          15         4

In [187]:
answer[answer['size']['neutral or dissatisfied'] < answer['size']['satisfied']]

size          
satisfaction neutral or dissatisfied satisfied
Class                                         
Business                          36        76

# 수질 음용성 여부 데이터

In [188]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/waters/train.csv")

In [189]:
df.head()

ph    Hardness        Solids  Chloramines     Sulfate  Conductivity  \
0  7.918150  214.186611  23823.492888     7.290878  341.173322    411.424483   
1  5.422446  205.266080  18542.957451     5.491963  306.702227    382.080129   
2  7.341547  187.672402  21273.457066     7.784003         NaN    332.084293   
3  9.056245  197.666301  17403.532167     7.688917  337.460176    414.766631   
4  5.039374  142.860598  40829.353167     7.271543         NaN    386.803057   

   Organic_carbon  Trihalomethanes  Turbidity  Potability  
0       19.585002        25.057375   4.028958           0  
1       10.504023        67.493450   2.911751           1  
2       16.842334        55.019151   4.025644           0  
3       15.349869        63.696746   3.319354           0  
4       16.823773        52.297113   4.957420           0

**Question 93. ph값은 상당히 많은 결측치를 포함한다. 결측치를 제외한 나머지 데이터들 중 사분위값 기준 하위 25%의 값들의 평균값은?**

In [190]:
target = df['ph'].dropna()
answer = target.loc[target <= target.quantile(0.25)].mean()
print(answer)

5.057093462441731


In [191]:
df[(df.ph.isnull()==False) & (df['ph']<=df['ph'].quantile(0.25))].ph.mean()

5.057093462441731

# 의료 비용 예측 데이터

In [192]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/MedicalCost/train.csv")

In [193]:
df.head()

age     sex     bmi  children smoker     region      charges
0   64  female  39.330         0     no  northeast  14901.51670
1   47  female  27.830         0    yes  southeast  23065.42070
2   52  female  33.300         2     no  southwest  10806.83900
3   33  female  22.135         1     no  northeast   5354.07465
4   30    male  31.400         1     no  southwest   3659.34600

**Question 94. 흡연자와 비흡연자 각각 charges의 상위 10% 그룹의 평균의 차이는?**

In [194]:
high = df.loc[df.smoker=='yes'].charges.quantile(0.9)
high2 = df.loc[df.smoker=='no'].charges.quantile(0.9)
mean_yes = df.loc[(df.smoker=='yes') & (df.charges>=high)].charges.mean()
mean_no = df.loc[(df.smoker=='no') & (df.charges>=high2)].charges.mean()
answer = mean_yes - mean_no
print(answer)

29297.954548156144


# 킹카운티 주거지 가격 예측 문제 데이터

In [195]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/kingcountyprice//train.csv")

In [196]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  8961990160  20150413T000000  567500.0         3        2.5         2080   
1  9455200205  20140604T000000  525000.0         3        2.0         1540   
2  7853220670  20140918T000000  540000.0         3        2.5         2860   
3  3298201170  20141110T000000  350000.0         3        1.0          940   
4  7972604355  20140521T000000  218000.0         3        1.0         1020   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      4556     2.0           0     0  ...      8        2080              0   
1      7800     1.0           0     0  ...      8        1540              0   
2      8935     2.0           0     0  ...      8        2860              0   
3      7811     1.0           0     0  ...      6         940              0   
4      7874     1.0           0     0  ...      7        1020              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1999             0    98074  47.6036 -122.014           1530   
1      2004             0    98125  47.7041 -122.288           1510   
2      2004             0    98065  47.5336 -121.855           2650   
3      1959             0    98008  47.6195 -122.118           1180   
4      1956             0    98106  47.5175 -122.346           1290   

   sqft_lot15  
0        5606  
1        7800  
2        6167  
3        7490  
4        7320  

[5 rows x 21 columns]

**Question 95. bedrooms 의 빈도가 가장 높은 값을 가지는 데이터들의 price의 상위 10%와 하위 10%값의 차이를 구하여라**

In [197]:
df.loc[df.bedrooms==df.bedrooms.value_counts().index[0]].price.quantile(0.9)-\
df.loc[df.bedrooms==df.bedrooms.value_counts().index[0]].price.quantile(0.1)

505500.0

# 대학원 입학 가능성 데이터

In [198]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/admission/train.csv")

In [199]:
df.head()

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
0         246        328          110                  4  4.0  2.5  9.02   
1         100        323          113                  3  4.0  4.0  8.88   
2          79        296           95                  2  3.0  2.0  7.54   
3          53        334          116                  4  4.0  3.0  8.00   
4         444        321          114                  5  4.5  4.5  9.16   

   Research  Chance of Admit  
0         1             0.81  
1         1             0.79  
2         1             0.44  
3         1             0.78  
4         1             0.87

**Question 96. Serial No. 컬럼을 제외하고 ‘Chance of Admit’을 종속변수, 나머지 변수를 독립변수라 할때, 랜덤포레스트를 통해 회귀 예측을 할 떄 변수중요도 값을 출력하라 (시드값에 따라 순서는 달라질수 있음)**

In [200]:
from sklearn.ensemble import RandomForestRegressor

In [201]:
x = df.iloc[:, 1:-1]
y = df['Chance of Admit']
rf = RandomForestRegressor()
model = rf.fit(x, y)
pd.DataFrame({'importance':model.feature_importances_}, x.columns).sort_values('importance', ascending=False)

importance
CGPA                 0.815047
GRE Score            0.086204
TOEFL Score          0.030773
LOR                  0.022734
SOP                  0.022287
University Rating    0.014614
Research             0.008342

# 레드 와인 퀄리티 예측 데이터

In [202]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/redwine/train.csv")

In [203]:
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            5.0              0.38         0.01             1.6      0.048   
1            5.0              0.42         0.24             2.0      0.060   
2            7.1              0.36         0.30             1.6      0.080   
3            7.6              0.29         0.49             2.7      0.092   
4            7.7              0.51         0.28             2.1      0.087   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 26.0                  60.0  0.99084  3.70       0.75   
1                 19.0                  50.0  0.99170  3.72       0.74   
2                 35.0                  70.0  0.99693  3.44       0.50   
3                 25.0                  60.0  0.99710  3.31       0.61   
4                 23.0                  54.0  0.99800  3.42       0.74   

   alcohol  quality  
0     14.0        6  
1     14.0        8  
2      9.4        5  
3     10.1        6  
4      9.2        5

**Question 97. quality 값이 3인 그룹과 8인 데이터그룹의 각 컬럼별 독립변수의 표준편차 값의 차이를 구할때 그값이 가장 큰 컬럼명을 구하여라**

In [204]:
(df.loc[df.quality ==8].std() -df.loc[df.quality ==3].std()).sort_values().index[-1]

'total sulfur dioxide'

In [205]:
abs(df[df['quality']==3].std() - df[df['quality']==8].std()).sort_values().index[-1]

'total sulfur dioxide'

# 약물 분류 데이터

In [206]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/drug/train.csv")

In [207]:
df.head()

Age Sex      BP Cholesterol  Na_to_K  Drug
0   20   F  NORMAL      NORMAL    9.281     4
1   24   M    HIGH      NORMAL    9.475     1
2   34   M  NORMAL        HIGH   22.456     0
3   17   M  NORMAL      NORMAL   10.832     4
4   72   M     LOW        HIGH   16.310     0

**Question 98. 남성들의 연령대별 Na_to_K값의 평균값을 구해서 데이터 프레임으로 표현하여라**

In [208]:
df['연령대'] = df['Age']//10*10
df[df['Sex']=='M'].groupby('연령대')['Na_to_K'].mean().to_frame()

Na_to_K
연령대           
10   13.627000
20   18.260769
30   15.244143
40   13.897273
50   14.811273
60   14.284308
70   11.363857

# 사기 회사 분류 데이터

In [209]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/audit/train.csv")

In [210]:
df.head()

Sector_score LOCATION_ID  PARA_A  Score_A  Risk_A  PARA_B  Score_B  Risk_B  \
0          2.72          14    4.53      0.6   2.718   87.93      0.6  52.758   
1          3.41          16   12.68      0.6   7.608   41.00      0.6  24.600   
2          2.72          11    3.11      0.6   1.866  113.97      0.6  68.382   
3          3.41           8    1.12      0.4   0.448    0.00      0.2   0.000   
4         55.57           9    1.06      0.4   0.424    0.00      0.2   0.000   

    TOTAL  numbers  ...  RiSk_E  History  Prob  Risk_F  Score  Inherent_Risk  \
0   92.46      5.0  ...     0.4        0   0.2     0.0    4.0        108.362   
1   53.68      5.5  ...     0.4        0   0.2     0.0    4.4         41.936   
2  117.08      5.5  ...     0.4        0   0.2     0.0    4.4         88.832   
3    1.12      5.0  ...     0.4        0   0.2     0.0    2.2          1.848   
4    1.06      5.0  ...     0.4        0   0.2     0.0    2.2          1.824   

   CONTROL_RISK  Detection_Risk  Audit_Risk  Risk  
0           0.4             0.5     21.6724     1  
1           0.4             0.5      8.3872     1  
2           0.4             0.5     17.7664     1  
3           0.4             0.5      0.3696     0  
4           0.4             0.5      0.3648     0  

[5 rows x 27 columns]

**Question 99. 데이터의 Risk 값에 따른 score_a와 score_b의 평균값을 구하여라**

In [211]:
df.groupby('Risk')[['Score_A', 'Score_B']].mean()

Score_A   Score_B
Risk                    
0     0.262234  0.225532
1     0.490164  0.444262

# 센서 데이터 동작 유형 분류 데이터

In [212]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/muscle/train.csv")

In [213]:
df.head()

motion_0  motion_1  motion_2  motion_3  motion_4  motion_5  motion_6  \
0      -6.0       8.0       5.0       6.0     -12.0     -27.0       5.0   
1     -32.0      -3.0      -4.0      -5.0      -1.0      16.0       0.0   
2      35.0      -2.0      -1.0       1.0     -25.0     -16.0      -3.0   
3     -15.0      -6.0      -3.0      -5.0      27.0      37.0      -2.0   
4      60.0       3.0       0.0      -7.0      -6.0     -10.0      -6.0   

   motion_7  motion_8  motion_9  ...  motion_55  motion_56  motion_57  \
0      24.0      -8.0      -5.0  ...       -1.0       -3.0       -6.0   
1      -5.0      42.0       2.0  ...      -17.0      -12.0       -1.0   
2     -20.0      10.0       1.0  ...        6.0      -65.0        6.0   
3      -2.0      19.0       5.0  ...        2.0       11.0        0.0   
4       7.0     -11.0      -4.0  ...      -12.0       17.0       -3.0   

   motion_58  motion_59  motion_60  motion_61  motion_62  motion_63  pose  
0       -9.0       -7.0       36.0       68.0        3.0       -3.0     1  
1       -1.0        0.0      -26.0      -11.0       -4.0       -9.0     1  
2      -10.0      -31.0      -19.0       -3.0      -47.0      -17.0     0  
3       -2.0       -4.0       -7.0       15.0        3.0       14.0     1  
4        1.0        2.0       -2.0       20.0       18.0       10.0     0  

[5 rows x 65 columns]

**Question 100. pose값에 따른 각 motion컬럼의 중간값의 가장 큰 차이를 보이는 motion컬럼은 어디이며 그 값은?**

In [214]:
t = df.groupby('pose').median().T
abs(t[0]-t[1]).sort_values()
display(t.tail(5))

pose         0    1
motion_59  0.0 -1.0
motion_60 -1.0  1.0
motion_61 -1.0  1.0
motion_62 -5.0 -1.0
motion_63 -1.0 -1.0

In [215]:
target = df.groupby('pose').median()
target = target.T
target['diff'] = abs(target[0] - target[1])
target[(target['diff']==target['diff'].max())]

pose         0    1  diff
motion_54 -5.0 -1.0   4.0
motion_62 -5.0 -1.0   4.0

# 현대 차량 가격 분류 문제 데이터

In [216]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/hyundai/train.csv")

In [217]:
df.head()

model  year  price transmission  mileage fuelType  tax(£)   mpg  \
0   Tucson  2016  12795       Manual    36263   Diesel      30  61.7   
1      I10  2012   3995       Manual    22089   Petrol      20  61.4   
2      I30  2019  24000       Manual      555   Petrol     145  34.0   
3   Tucson  2017  12995       Manual    32147   Diesel      30  61.7   
4   Tucson  2018  14991    Semi-Auto    32217   Diesel     150  57.6   

   engineSize  
0         1.7  
1         1.2  
2         2.0  
3         1.7  
4         1.7

**Question 101. 정보(row수)가 가장 많은 상위 3차종의 price값의 각 평균값은?**

In [218]:
df.loc[df.model.isin(df.model.value_counts().index[:3])].groupby('model')['price'].mean().to_frame()

price
model                
 I10      7646.137891
 I30     11651.821759
 Tucson  15805.441373

# 당뇨 여부 판단 데이터

In [219]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/diabetes/train.csv")

In [220]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            3      102             74              0        0  29.5   
1            2      144             58             33      135  31.6   
2            5      136             82              0        0   0.0   
3           13      145             82             19      110  22.2   
4            1      117             60             23      106  33.8   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.121   32        0  
1                     0.422   25        1  
2                     0.640   69        0  
3                     0.245   57        0  
4                     0.466   27        0

**Question 102. Outcome 값에 따른 각 그룹의 각 컬럼의 평균 차이를 구하여라**

In [221]:
df.groupby('Outcome').mean().diff().iloc[1, :]

Pregnancies                  1.574159
Glucose                     29.943995
BloodPressure                3.202079
SkinThickness                2.952033
Insulin                     33.398645
BMI                          4.541437
DiabetesPedigreeFunction     0.143646
Age                          5.866939
Name: 1, dtype: float64

# 넷플릭스 주식 데이터

In [222]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nflx/NFLX.csv")

In [223]:
df.head()

Date      Open      High       Low     Close  Adj Close     Volume
0  2002-05-23  1.156429  1.242857  1.145714  1.196429   1.196429  104790000
1  2002-05-24  1.214286  1.225000  1.197143  1.210000   1.210000   11104800
2  2002-05-28  1.213571  1.232143  1.157143  1.157143   1.157143    6609400
3  2002-05-29  1.164286  1.164286  1.085714  1.103571   1.103571    6757800
4  2002-05-30  1.107857  1.107857  1.071429  1.071429   1.071429   10154200

**Question 103. 매년 5월달의 open가격의 평균값을 데이터 프레임으로 표현하라**

In [224]:
df['Date'] = pd.to_datetime(df['Date'])
df.groupby(df[df['Date'].dt.month==5]['Date'].dt.strftime('%Y-%m'))['Open'].mean()

Date
2002-05      1.155833
2003-05      1.641497
2004-05      4.261143
2005-05      1.951905
2006-05      4.163571
2007-05      3.159351
2008-05      4.435034
2009-05      5.790571
2010-05     14.417071
2011-05     34.650272
2012-05     10.523247
2013-05     31.936429
2014-05     51.121292
2015-05     85.057429
2016-05     92.705715
2017-05    158.255455
2018-05    329.779541
2019-05    359.664548
2020-05    433.880499
2021-05    496.923996
Name: Open, dtype: float64

In [225]:
target = df.groupby(df['Date'].dt.strftime('%Y-%m')).mean(numeric_only=True)
answer = target.loc[target.index.str.contains('-05')].Open
print(answer)

Date
2002-05      1.155833
2003-05      1.641497
2004-05      4.261143
2005-05      1.951905
2006-05      4.163571
2007-05      3.159351
2008-05      4.435034
2009-05      5.790571
2010-05     14.417071
2011-05     34.650272
2012-05     10.523247
2013-05     31.936429
2014-05     51.121292
2015-05     85.057429
2016-05     92.705715
2017-05    158.255455
2018-05    329.779541
2019-05    359.664548
2020-05    433.880499
2021-05    496.923996
Name: Open, dtype: float64


# NBA 선수 능력치 데이터

In [226]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nba/nba.csv", encoding='latin', sep=';')

In [227]:
df.head()

Rk             Player Pos  Age   Tm   G  GS    MP   FG   FGA  ...    FT%  \
0   1   Precious Achiuwa   C   22  TOR  73  28  23.6  3.6   8.3  ...  0.595   
1   2       Steven Adams   C   28  MEM  76  75  26.3  2.8   5.1  ...  0.543   
2   3        Bam Adebayo   C   24  MIA  56  56  32.6  7.3  13.0  ...  0.753   
3   4       Santi Aldama  PF   21  MEM  32   0  11.3  1.7   4.1  ...  0.625   
4   5  LaMarcus Aldridge   C   36  BRK  47  12  22.3  5.4   9.7  ...  0.873   

   ORB  DRB   TRB  AST  STL  BLK  TOV   PF   PTS  
0  2.0  4.5   6.5  1.1  0.5  0.6  1.2  2.1   9.1  
1  4.6  5.4  10.0  3.4  0.9  0.8  1.5  2.0   6.9  
2  2.4  7.6  10.1  3.4  1.4  0.8  2.6  3.1  19.1  
3  1.0  1.7   2.7  0.7  0.2  0.3  0.5  1.1   4.1  
4  1.6  3.9   5.5  0.9  0.3  1.0  0.9  1.7  12.9  

[5 rows x 30 columns]

**Question 104. Tm 컬럼은 각 팀의 이름을 의미한다. TOR팀의 평균 나이를 소수 4째 자리까지 구하여라**

In [228]:
round(df[df['Tm']=='TOR']['Age'].mean(), 4)

24.8696

**Question 105. Pos 컬럼은 포지션을 의미한다. 전체 선수 중 최소 나이대의 선수들을 필터하고 그들 중 가장 많은 포지션은 무엇인지 확인하라**

In [229]:
df[df.Age==df.Age.min()].Pos.value_counts().index[0]

'SG'

In [230]:
df[df['Age']==19].groupby('Pos').size().sort_values().index[-1]

'SG'

**Question 106. 선수들의 이름은 first_name+ 공백 + last_name으로 이루어져 있다. 가장 많은 first_name은 무엇이며 몇 회 발생하는지 확인하라**

In [231]:
df['Player'].str.split(' ').str[0].str.lower().value_counts().head(1)

justin    14
Name: Player, dtype: int64

**Question 107. PTS컬럼은 경기당 평균득점수이다. 각 포지션별로 경기당 평균득점수의 평균을 구하여라**

In [232]:
df.groupby('Pos')['PTS'].mean().sort_values()

Pos
SF        7.270253
SG-SF     7.360000
C         7.690769
PF        7.737500
SG        7.810553
PG        8.325161
PF-SF     8.500000
SF-SG     8.660000
SG-PG     9.525000
C-PF     12.850000
PG-SG    22.000000
Name: PTS, dtype: float64

**Question 108. G컬럼은 참여한 경기의 숫자이다. 각 팀별로 가장 높은 경기참여 수를 가진 선수들의 경기 참여 숫자의 평균을 구하여라**

In [233]:
df.sort_values(['Tm', 'G']).groupby('Tm').tail(1).G.mean()

77.51612903225806

In [234]:
df.sort_values(['Tm', 'G']).groupby('Tm').tail()

Rk                    Player Pos  Age   Tm   G  GS    MP   FG   FGA  \
242  182          Danilo Gallinari  PF   33  ATL  66  18  25.3  3.9   9.0   
112   93              Clint Capela   C   27  ATL  74  73  27.6  5.0   8.2   
341  255             Kevin Huerter  SG   23  ATL  74  60  29.6  4.7  10.3   
808  602                Trae Young  PG   23  ATL  76  76  34.9  9.4  20.3   
799  597              Delon Wright  SG   29  ATL  77   8  18.9  1.6   3.5   
..   ...                       ...  ..  ...  ...  ..  ..   ...  ...   ...   
549  407                 Raul Neto  PG   29  WAS  70  19  19.6  2.9   6.3   
241  181            Daniel Gafford   C   23  WAS  72  53  20.1  4.0   5.7   
108   89  Kentavious Caldwell-Pope  SG   28  WAS  77  77  30.2  4.7  10.8   
417  308             Corey Kispert  SF   22  WAS  77  36  23.4  3.0   6.7   
25    20               Deni Avdija  SF   21  WAS  82   8  24.2  3.0   7.1   

     ...    FT%  ORB  DRB   TRB  AST  STL  BLK  TOV   PF   PTS  
242  ...  0.904  0.5  4.1   4.7  1.5  0.4  0.2  0.6  1.4  11.7  
112  ...  0.473  3.8  8.1  11.9  1.2  0.7  1.3  0.6  2.2  11.1  
341  ...  0.808  0.4  3.0   3.4  2.7  0.7  0.4  1.2  2.5  12.1  
808  ...  0.904  0.7  3.1   3.7  9.7  0.9  0.1  4.0  1.7  28.4  
799  ...  0.857  0.6  2.2   2.9  2.4  1.2  0.2  0.6  0.7   4.4  
..   ...    ...  ...  ...   ...  ...  ...  ...  ...  ...   ...  
549  ...  0.769  0.2  1.7   1.9  3.1  0.8  0.0  1.1  1.5   7.5  
241  ...  0.699  2.2  3.5   5.7  0.9  0.4  1.4  0.9  2.4   9.4  
108  ...  0.890  0.5  2.9   3.4  1.9  1.1  0.3  1.3  1.8  13.2  
417  ...  0.871  0.7  2.0   2.7  1.1  0.5  0.3  0.6  1.2   8.2  
25   ...  0.757  0.6  4.5   5.2  2.0  0.7  0.5  1.1  2.3   8.4  

[155 rows x 30 columns]

**Question 110. Tm의 값이 MIA이며 Pos는 C또는 PF인 선수의 MP값의 평균은?**

In [235]:
df[(df.Tm=='MIA') & (df.Pos.isin(['C','PF']))].MP.mean()

16.7875

In [236]:
df[(df['Tm']=='MIA') & ((df['Pos']=='C') | (df['Pos']=='PF'))]['MP'].mean()

16.7875

**Question 111. 전체 데이터중 G의 평균값의 1.5배 이상인 데이터들만 추출했을 때 3P값의 평균은?**

In [237]:
df[df['G']>=(df['G'].mean()*1.5)]['3P'].mean()

1.3853658536585365

**Question 112. Age의 평균 이상인 그룹과 평균 미만인 그룹간의 G값의 평균의 차이는?**

In [238]:
df[df.Age >= df.Age.mean()].G.mean() - df[df.Age < df.Age.mean()].G.mean()

3.787674551781862

In [239]:
over = df[df['Age']>=df['Age'].mean()]
under = df[df['Age']<df['Age'].mean()]

over.G.mean() - under.G.mean()

3.787674551781862

**Question 113. 평균 나이가 가장 젊은 팀은 어디인가**

In [240]:
df.groupby('Tm')['Age'].mean().sort_values().index[0]

'MEM'

**Question 114. Pos그룹별 평균 MP값을 구하여라**

In [241]:
df.groupby('Pos')['MP'].mean()

Pos
C        16.990000
C-PF     25.350000
PF       17.937500
PF-SF    27.300000
PG       19.547742
PG-SG    37.200000
SF       17.514557
SF-SG    20.340000
SG       18.554271
SG-PG    22.950000
SG-SF    19.620000
Name: MP, dtype: float64